<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Panomics Marketplace - Quality Control and Statistical Analysis for Panomics Data • pmartR</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="bootstrap-toc.css">
<script src="bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="pkgdown.css" rel="stylesheet">
<script src="pkgdown.js"></script><meta property="og:title" content="Panomics Marketplace - Quality Control and Statistical Analysis for Panomics Data">
<meta property="og:description" content="Provides functionality for quality control processing and statistical analysis of mass spectrometry (MS) omics data, in particular proteomic (either at the peptide or the protein level), lipidomic, and metabolomic data, as well as RNA-seq based count data and nuclear magnetic resonance (NMR) data. This includes data transformation, specification of groups that are to be compared against each other, filtering of features and/or samples, data normalization, data summarization (correlation, PCA), and statistical comparisons between defined groups.  Implements methods described in:  Webb-Robertson et al. (2014) &lt;doi:10.1074/mcp.M113.030932&gt;.  Webb-Robertson et al. (2011) &lt;doi:10.1002/pmic.201100078&gt;.  Matzke et al. (2011) &lt;doi:10.1093/bioinformatics/btr479&gt;.  Matzke et al. (2013) &lt;doi:10.1002/pmic.201200269&gt;.  Polpitiya et al. (2008) &lt;doi:10.1093/bioinformatics/btn217&gt;.  Webb-Robertson et al. (2010) &lt;doi:10.1021/pr1005247&gt;.  ">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-home">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="index.html">pmartR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.4.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="articles/Convenience_Functions.html">Convenience Functions</a>
    </li>
    <li>
      <a href="articles/Data_Normalization.html">Data Normalization</a>
    </li>
    <li>
      <a href="articles/Data_Object_Creation.html">Data Object Creation</a>
    </li>
    <li>
      <a href="articles/Filter_Functionality.html">Filter Functions</a>
    </li>
    <li>
      <a href="articles/Graphing_with_pmartR.html">Graphing with pmartR</a>
    </li>
    <li>
      <a href="articles/RNAseq_Processing_Workflow.html">RNAseq Data Processing</a>
    </li>
    <li>
      <a href="articles/Statistical_Analysis_with_pmartR.html">Statistical Analysis with the `pmartR` Package</a>
    </li>
    <li>
      <a href="articles/Trelliscope_Vignette.html">Using Trelliscope Functions with pmartR</a>
    </li>
    <li>
      <a href="articles/Typical_Processing_Workflow.html">Typical Processing Workflow</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/pmartR/pmartR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="contents col-md-9">
<div class="section level1">
<div class="page-header"><h1 id="pmartr">pmartR<a class="anchor" aria-label="anchor" href="#pmartr"></a>
</h1></div>
<!-- badges: start -->

<p>This R package provides functionality for quality control processing, statistical analysis and visualization of mass spectrometry (MS) omics data, in particular proteomic (either at the peptide or the protein level; isobaric labeled or unlabled), lipidomic, and metabolomic data. This includes data transformation, specification of groups that are to be compared against each other, filtering of feature and/or samples, data normalization, data summarization (correlation, PCA), and statistical comparisons of groups of interest (ANOVA and/or independence of missing data tests). Example data to be used with this packages can be found in <a href="https://github.com/pmartR/pmartRdata" class="external-link">pmartRdata</a>.</p>
<div class="section level2">
<h2 id="installation">Installation:<a class="anchor" aria-label="anchor" href="#installation"></a>
</h2>
<p>This package makes use of several packages hosted on BioConductor. If you are encountering warnings about unavailable BioConductor packages such as <code>pcaMethods</code>, you may need to add them to <code>options("repos")</code>:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/ns-load.html" class="external-link">requireNamespace</a></span><span class="op">(</span><span class="st">"BiocManager"</span>, quietly <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"BiocManager"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/options.html" class="external-link">options</a></span><span class="op">(</span><span class="st">"repos"</span> <span class="op">=</span> <span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">repositories</span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>(<strong>Recommended</strong>) Install from CRAN:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/install.packages.html" class="external-link">install.packages</a></span><span class="op">(</span><span class="st">"pmartR"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># or </span></span>
<span></span>
<span><span class="fu">BiocManager</span><span class="fu">::</span><span class="fu">install</span><span class="op">(</span><span class="st">"pmartR"</span><span class="op">)</span></span></code></pre></div>
<p>To install the latest <em>release</em>:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"pmartR/pmartR@*release"</span><span class="op">)</span></span></code></pre></div>
<p>To install a specific release, say v2.4.0:</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"pmartR/pmartR@v2.4.0"</span><span class="op">)</span></span></code></pre></div>
<p><strong>(Not recommended, since these changes are likely still being tested)</strong> You can also install the latest changes to master:</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">devtools</span><span class="fu">::</span><span class="fu">install_github</span><span class="op">(</span><span class="st">"pmartR/pmartR"</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="problems-with-rcpparmadillo-and-gfortran-on-mac">Problems with rcppArmadillo and gfortran on mac<a class="anchor" aria-label="anchor" href="#problems-with-rcpparmadillo-and-gfortran-on-mac"></a>
</h3>
<p>There is a problem that causes pmartR to fail compiling cpp code, which has something to do with rcppArmadillo and certain installations of gfortran. See these posts that try to explain the issue: <a href="https://stackoverflow.com/questions/64992467/mac-clang-installation-seems-to-override-gcc-install" class="external-link">1</a> <a href="https://stackoverflow.com/questions/29992066/rcpp-warning-directory-not-found-for-option-l-usr-local-cellar-gfortran-4-8/29993906#29993906" class="external-link">2</a> <a href="https://community.rstudio.com/t/setting-up-travis-ci-on-linux-with-an-r-package-that-uses-rcpparmadillo/53910/3" class="external-link">3</a>. Two solutions we have found:</p>
<ol style="list-style-type: decimal">
<li>Install gfortran from a recommended source (not homebrew):
<ul>
<li>
<a href="https://mac.r-project.org/tools/" class="external-link">This CRAN-approved resource for build tools on mac</a> lists two versions of gfortran and how to install them.</li>
<li>On Catalina 10.15.7 I downloaded and installed gfortran 8.2 from the link provided in <a href="https://thecoatlessprofessor.com/programming/cpp/r-compiler-tools-for-rcpp-on-macos/#google_vignette" class="external-link">this blog post</a><br>
</li>
</ul>
</li>
<li>When using the homebrew gfortran installation, add the line <strong>FLIBS = -L`gfortran -print-file-name=libgfortran.dylib | xargs dirname`</strong> to ~/.R/Makevars (a plain text file with no extention)</li>
</ol>
</div>
<div class="section level3">
<h3 id="gfortran-and-apple-silicon-m1m2-chips">gfortran and Apple silicon (M1/M2 chips)<a class="anchor" aria-label="anchor" href="#gfortran-and-apple-silicon-m1m2-chips"></a>
</h3>
<p>There are similarly issues with compilation in newer Mac chips. We recommend to install gcc-13 from homebrew <code>brew install gcc</code> or the universal version from <a href="https://mac.r-project.org/tools/" class="external-link uri">https://mac.r-project.org/tools/</a>.</p>
<p>Additionally, some users experience errors with <code>ld:  Assertion failed ...</code> as seen <a href="https://developer.apple.com/forums/thread/737707" class="external-link">here</a>. One solution is to use the old linker by making sure gcc uses the flag <code>-ld64</code> <a href="https://developer.apple.com/documentation/xcode-release-notes/xcode-15-release-notes#Linking" class="external-link">(Xcode docs)</a>. To do this, you can edit <code>~/.R/Makevars</code> to include this flag, for example by appending it to LDFLAGS with +=:</p>
<pre><code># in ~/.R/Makevars
LDFLAGS+=-ld64</code></pre>
<p>or specifying it in your compiler command:</p>
<pre><code># in ~/.R/Makevars
CC=/usr/local/bin/gcc -ld64</code></pre>
</div>
</div>
<div class="section level2">
<h2 id="tutorial">Tutorial:<a class="anchor" aria-label="anchor" href="#tutorial"></a>
</h2>
<p>To get started, see the package documentation and function reference located <a href="https://pmartr.github.io/pmartR/" class="external-link">here</a>.</p>
</div>
<div class="section level2">
<h2 id="data">Data:<a class="anchor" aria-label="anchor" href="#data"></a>
</h2>
<p>Example peptide (both unlabeled and isobaric labeled), protein, metabolite and lipid data are available in the <strong>pmartRdata</strong> package available on Github, <a href="https://github.com/pmartR/pmartRdata" class="external-link">here</a></p>
</div>
<div class="section level2">
<h2 id="citation">Citation:<a class="anchor" aria-label="anchor" href="#citation"></a>
</h2>
<p>To cite this package, please the following:</p>
<p>Degnan, D. J.; Stratton, K. G.; Richardson, R.; Claborne, D.; Martin, E. A.; Johnson, N. A.; Leach, D.; Webb-Robertson, B.-J. M.; Bramer, L. M. PmartR 2.0: A Quality Control, Visualization, and Statistics Pipeline for Multiple Omics Datatypes. J. Proteome Res. 2023, 22 (2), 570–576. <a href="https://doi.org/10.1021/acs.jproteome.2c00610" class="external-link uri">https://doi.org/10.1021/acs.jproteome.2c00610</a>.</p>
<p><strong>BibTex:</strong></p>
<pre><code>@article{degnan2023pmartr,
  title={pmartR 2.0: A Quality Control, Visualization, and Statistics Pipeline for Multiple Omics Datatypes},
  author={Degnan, David J and Stratton, Kelly G and Richardson, Rachel and Claborne, Daniel and Martin, Evan A and Johnson, Nathan A and Leach, Damon and Webb-Robertson, Bobbie-Jo M and Bramer, Lisa M},
  doi = {10.1021/acs.jproteome.2c00610},
  journal={Journal of Proteome Research},
  year={2023},
  publisher={ACS Publications}
}) </code></pre>
</div>
<div class="section level2">
<h2 id="disclaimer">Disclaimer:<a class="anchor" aria-label="anchor" href="#disclaimer"></a>
</h2>
<p>This material was prepared as an account of work sponsored by an agency of the United States Government. Neither the United States Government nor the United States Department of Energy, nor Battelle, nor any of their employees, nor any jurisdiction or organization that has cooperated in the development of these materials, makes any warranty, express or implied, or assumes any legal liability or responsibility for the accuracy, completeness, or usefulness or any information, apparatus, product, software, or process disclosed, or represents that its use would not infringe privately owned rights.</p>
<p>Reference herein to any specific commercial product, process, or service by trade name, trademark, manufacturer, or otherwise does not necessarily constitute or imply its endorsement, recommendation, or favoring by the United States Government or any agency thereof, or Battelle Memorial Institute. The views and opinions of authors expressed herein do not necessarily state or reflect those of the United States Government or any agency thereof.</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode R"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>  PACIFIC NORTHWEST NATIONAL LABORATORY</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  operated by BATTELLE <span class="cf">for</span> the</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  UNITED STATES DEPARTMENT OF ENERGY</span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>  under Contract DE<span class="sc">-</span>AC05<span class="sc">-</span>76RL01830</span></code></pre></div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <div class="links">
<h2 data-toc-skip>Links</h2>
<ul class="list-unstyled">
<li><a href="https://cloud.r-project.org/package=pmartR" class="external-link">View on CRAN</a></li>
<li><a href="https://github.com/pmartR/pmartR/" class="external-link">Browse source code</a></li>
<li><a href="https://github.com/pmartR/pmartR/issues" class="external-link">Report a bug</a></li>
</ul>
</div>

<div class="license">
<h2 data-toc-skip>License</h2>
<ul class="list-unstyled">
<li><a href="LICENSE.html">Full license</a></li>
<li><small><a href="https://opensource.org/licenses/BSD-2-Clause" class="external-link">BSD_2_clause</a> + file <a href="LICENSE-text.html">LICENSE</a></small></li>
</ul>
</div>


<div class="citation">
<h2 data-toc-skip>Citation</h2>
<ul class="list-unstyled">
<li><a href="authors.html#citation">Citing pmartR</a></li>
</ul>
</div>

<div class="developers">
<h2 data-toc-skip>Developers</h2>
<ul class="list-unstyled">
<li>Lisa Bramer <br><small class="roles"> Author, maintainer </small>  </li>
<li>Kelly Stratton <br><small class="roles"> Author </small>  </li>
<li>Daniel Claborne <br><small class="roles"> Author </small>  </li>
<li><a href="authors.html">More about authors...</a></li>
</ul>
</div>

<div class="dev-status">
<h2 data-toc-skip>Dev status</h2>
<ul class="list-unstyled">
<li><a href="https://zenodo.org/badge/latestdoi/69275428" class="external-link"><img src="https://zenodo.org/badge/69275428.svg" alt="DOI"></a></li>
<li><a href="https://github.com/pmartR/pmartR/actions/workflows/R-CMD-check.yaml" class="external-link"><img src="https://github.com/pmartR/pmartR/actions/workflows/R-CMD-check.yaml/badge.svg" alt="R-CMD-check"></a></li>
<li><a href="https://CRAN.R-project.org/package=pmartR" class="external-link"><img src="https://www.r-pkg.org/badges/version/pmartR" alt="CRAN status"></a></li>
</ul>
</div>

  </div>
</div>


      <footer><div class="copyright">
  <p></p>
<p>Developed by Lisa Bramer, Kelly Stratton, Daniel Claborne.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
