---
title: "RNAseq Data Processing"
author: "Rachel Richardson, Kelly Stratton, Lisa Bramer"
date: "`r Sys.Date()`"
output: 
  html_document:
    df_print: paged
    rmarkdown::html_vignette:
    fig_caption: yes
    self_contained: yes
    toc: true
bibliography: bibliography.bib
vignette: >
  %\VignetteIndexEntry{pmartR-RNAseq} 
  %\VignetteEngine{knitr::rmarkdown} 
  %\VignetteEncoding{UTF-8}
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
options(rmarkdown.html_vignette.check_title = FALSE)
```

## Overview

RNAseq data in `pmartR` follows a different workflow compared to the other data types, with different filtering, plot, and statistical options. This vignette walks through the RNAseq workflow, including the use of all three statistical analysis options, DEseq2 [@deseq2], edgeR [@edger], or limmaVOOM [@limmavoom].


## Data Set-up

To create a `seqData` object, an expression data frame (of count data) is required, as well as a data frame containing sample information. Biomolecule metadata is optional. Below we create a `seqData` object from example data frames in the `pmartRdata` package.

```{r}
library(pmartR)
library(pmartRdata)

data(rnaseq_edata)
data(rnaseq_fdata)
data(rnaseq_emeta)

mydata <- as.seqData(e_data = rnaseq_edata, f_data = rnaseq_fdata, e_meta = rnaseq_emeta,
                     edata_cname = "Transcript", fdata_cname = "SampleName", emeta_cname = "Gene",
                     data_scale = "counts", check.names = FALSE)
```

We can assign groups of interest for downstream statistical comparisons.

```{r}
mydata <- group_designation(mydata, main_effects = c("Virus"))
```



Unlike the other data types in `pmartR`, seqData objects are not transformed e.g. to a log scale as part of the standard workflow. In order to aid in the visualization of the data, the `plot()` function has an optional argument to transform the count data to one of the following:

- lcpm = log counts per million

- upper = 75th quantile scaling

- median = median scaling

```{r}
plot(mydata, transformation = "lcpm", order_by = "Virus", color_by = "Virus")
plot(mydata, transformation = "upper", order_by = "Virus", color_by = "Virus")
plot(mydata, transformation = "median", order_by = "Virus", color_by = "Virus")
```



## Quality Control

### Filters

Many of the available filters are specific to sequence (count) data:

- Total count filter

- RNA_filter = filters sample by total library size (number of reads) or by the number of unique non-zero transcripts per sample

- number of nonzero reads for each sample

The molecule_filter and custom_filter functions are available for seqData objects, however the cv_filter, imdanova_filter, and rmd_filter do not work for seqData objects.



```{r}
# total count filter
total_count_filter(mydata)

plot(total_count_filter(mydata))
plot(total_count_filter(mydata), min_count = 50)
plot(total_count_filter(mydata), palette = "Blues", min_count = 20)
plot(total_count_filter(mydata), palette = "Blues", 
     min_count = 20, interactive = T)
```

```{r}
# RNA filter
RNA_filter(mydata)

plot(RNA_filter(mydata))
plot(RNA_filter(mydata), size_library = 10000)
plot(RNA_filter(mydata), plot_type = "biomolecule")
plot(RNA_filter(mydata), min_nonzero = 700, plot_type = "biomolecule")
```



## Exploratory Data Analysis & Plots

```{r}
corres <- cor_result(mydata)
dimred <- dim_reduction(mydata, k = 2)

plot(corres)
plot(dimred)
```


## Normalization & Statistical Comparisons

Stats/normalization

Give guidance/table with pros cons of the 3 methods
User picks a method, then we show them the dispersion plot


```{r}
### Dispersion estimate wrapper

dispersion_est(mydata, "DESeq2")
dispersion_est(mydata, "edgeR")
dispersion_est(mydata, "voom")

```


User says "ok looks fine, given me the stat results" or "nope. no good, I'll try another method"

Plot of stat results - option for either volcano or MA plot, heat maps with option to subset to most significant molecules (based on p-value)

```{r}

deseq_res <- Deseq2_wrapper(mydata)
plot(deseq_res, plot_type = "bar")
plot(deseq_res, plot_type = "volcano")
plot(deseq_res, plot_type = "MA")

testthat::expect_error(plot(deseq_res, plot_type = "gheatmap"), "Not valid for seqData")

```


```{r}

edger_res <- EdgeR_wrapper(mydata)

plot(edger_res, plot_type = "bar") 
plot(edger_res, plot_type = "volcano")
plot(edger_res, plot_type = "MA")
testthat::expect_error(plot(edger_res, plot_type = "gheatmap"))

```


```{r}

voom_res <- Voom_wrapper(mydata)

plot(voom_res, plot_type = "bar")
plot(voom_res, plot_type = "volcano")
plot(voom_res, plot_type = "MA")
testthat::expect_error(plot(voom_res, plot_type = "gheatmap"), "Not valid for seqData")

```




## References

