<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Tests for a quantiative difference between groups (aka factors, aka main
effects) — anova_test • pmartR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Tests for a quantiative difference between groups (aka factors, aka main
effects) — anova_test"><meta property="og:description" content="This is the ANOVA part of the IMD-ANOVA test proposed in Webb-Robertson et
al. (2010)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pmartR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Convenience_Functions.html">Convenience Functions</a>
    </li>
    <li>
      <a href="../articles/Data_Normalization.html">Data Normalization</a>
    </li>
    <li>
      <a href="../articles/Data_Object_Creation.html">Data Object Creation</a>
    </li>
    <li>
      <a href="../articles/Filter_Functionality.html">Filter Functions</a>
    </li>
    <li>
      <a href="../articles/Graphing_with_pmartR.html">Graphing with pmartR</a>
    </li>
    <li>
      <a href="../articles/RNAseq_Processing_Workflow.html">RNAseq Data Processing</a>
    </li>
    <li>
      <a href="../articles/Statistical_Analysis_with_pmartR.html">Statistical Analysis with the `pmartR` Package</a>
    </li>
    <li>
      <a href="../articles/Trelliscope_Vignette.html">Using Trelliscope Functions with pmartR</a>
    </li>
    <li>
      <a href="../articles/Typical_Processing_Workflow.html">Typical Processing Workflow</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/pmartR/pmartR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Tests for a quantiative difference between groups (aka factors, aka main
effects)</h1>
    <small class="dont-index">Source: <a href="https://github.com/pmartR/pmartR/blob/HEAD/R/imd_anova.R" class="external-link"><code>R/imd_anova.R</code></a></small>
    <div class="hidden name"><code>anova_test.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This is the ANOVA part of the IMD-ANOVA test proposed in Webb-Robertson et
al. (2010).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">anova_test</span><span class="op">(</span></span>
<span>  <span class="va">omicsData</span>,</span>
<span>  <span class="va">groupData</span>,</span>
<span>  <span class="va">comparisons</span>,</span>
<span>  <span class="va">pval_adjust_multcomp</span>,</span>
<span>  <span class="va">pval_adjust_fdr</span>,</span>
<span>  <span class="va">pval_thresh</span>,</span>
<span>  <span class="va">covariates</span>,</span>
<span>  <span class="va">paired</span>,</span>
<span>  <span class="va">equal_var</span>,</span>
<span>  <span class="va">parallel</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>omicsData</dt>
<dd><p>A pmartR data object of any class</p></dd>


<dt>groupData</dt>
<dd><p>`data.frame` that assigns sample names to groups</p></dd>


<dt>comparisons</dt>
<dd><p>`data.frame` with columns for "Control" and "Test"
containing the different comparisons of interest. Comparisons will be made
between the Test and the corresponding Control  If left NULL, then all
pairwise comparisons are executed.</p></dd>


<dt>pval_adjust_multcomp</dt>
<dd><p>character string specifying the type of multiple
comparisons adjustment to implement. The default, "none", corresponds to no
adjustment. Valid options include: "bonferroni", "holm", "tukey", and
"dunnett".</p></dd>


<dt>pval_adjust_fdr</dt>
<dd><p>character string specifying the type of FDR adjustment
to implement. The default, "none", corresponds to no adjustment. Valid
options include: "bonferroni", "BH", "BY", and "fdr".</p></dd>


<dt>pval_thresh</dt>
<dd><p>numeric p-value threshold, below or equal to which
peptides are considered differentially expressed. Defaults to 0.05</p></dd>


<dt>covariates</dt>
<dd><p>A character vector with no more than two variable names
that will be used as covariates in the IMD-ANOVA analysis.</p></dd>


<dt>paired</dt>
<dd><p>logical; should the data be paired or not? if TRUE then the
`f_data` element of `omicsData` is checked for a "Pair" column, an error is
returned if none is found</p></dd>


<dt>equal_var</dt>
<dd><p>logical; should the variance across groups be assumed equal?</p></dd>


<dt>parallel</dt>
<dd><p>A logical value indicating if the t test should be run in
parallel.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>a list of `data.frame`s</p>
<p></p>
<table class="table table"><tr><td>Results</td><td>Edata cname,
  Variance Estimate, ANOVA F-Statistic, ANOVA p-value, Group means</td></tr><tr><td></td><td></td></tr><tr><td>Fold_changes</td><td>Estimated fold-changes for each comparison</td></tr><tr><td></td><td></td></tr><tr><td>Fold_changes_pvalues</td><td>P-values corresponding to the fold-changes
  for each comparison</td></tr><tr><td></td><td></td></tr><tr><td>Fold_change_flags</td><td>Indicator of
  statistical significance (0/+-2 to if adjusted p-value&gt;=pval_thresh or
  p-value&lt;pval_thresh)</td></tr></table></div>
    <div id="details">
    <h2>Details</h2>
    <p>The order in which different scenarios are handeled:</p><ol><li><p>If the
data are paired, then the pairing is accounted for first then each of the
next steps is carried out on the new variable that is the difference in the
paired individuals.&lt;br&gt;</p></li>
<li><p>If covariates are provided, their effect is
removed before testing for group differences though mathematically covariates
and grouping effects are accounted for simultaneously</p></li>
<li><p>ANOVA is executed
to assess the effect of each main effects, results in a vector of group means
for each biomolecule and variance estimate</p></li>
<li><p>Group comparisons defined by
`comaprison` argument are implemented use parameter vector and variance
estimates in ANOVA step</p></li>
</ol></div>
    <div id="references">
    <h2>References</h2>
    <p>Webb-Robertson, Bobbie-Jo M., et al. "Combined statistical
  analyses of peptide intensities and peptide occurrences improves
  identification of significant peptides from MS-based proteomics data."
  Journal of proteome research 9.11 (2010): 5748-5756.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Bryan Stanfill</p>
    </div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Lisa Bramer, Kelly Stratton.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

