<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Create pmartR Object of Class seqData — as.seqData • pmartR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Create pmartR Object of Class seqData — as.seqData"><meta property="og:description" content="Converts several data frames of RNA-seq transcript data to
an object of the class 'seqData'. Objects of the class 'seqData' are lists
with two obligatory components, e_data and f_data. An optional
list component, e_meta, is used if analysis or visualization at other
levels (e.g. gene, protein, pathway) is also desired."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pmartR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.4.5</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Convenience_Functions.html">Convenience Functions</a>
    </li>
    <li>
      <a href="../articles/Data_Normalization.html">Data Normalization</a>
    </li>
    <li>
      <a href="../articles/Data_Object_Creation.html">Data Object Creation</a>
    </li>
    <li>
      <a href="../articles/Filter_Functionality.html">Filter Functions</a>
    </li>
    <li>
      <a href="../articles/Graphing_with_pmartR.html">Graphing with pmartR</a>
    </li>
    <li>
      <a href="../articles/RNAseq_Processing_Workflow.html">RNAseq Data Processing</a>
    </li>
    <li>
      <a href="../articles/Statistical_Analysis_with_pmartR.html">Statistical Analysis with the `pmartR` Package</a>
    </li>
    <li>
      <a href="../articles/Trelliscope_Vignette.html">Using Trelliscope Functions with pmartR</a>
    </li>
    <li>
      <a href="../articles/Typical_Processing_Workflow.html">Typical Processing Workflow</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/pmartR/pmartR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Create pmartR Object of Class seqData</h1>
    <small class="dont-index">Source: <a href="https://github.com/pmartR/pmartR/blob/HEAD/R/as.omicsData.R" class="external-link"><code>R/as.omicsData.R</code></a></small>
    <div class="hidden name"><code>as.seqData.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Converts several data frames of RNA-seq transcript data to
an object of the class 'seqData'. Objects of the class 'seqData' are lists
with two obligatory components, <code>e_data</code> and <code>f_data</code>. An optional
list component, <code>e_meta</code>, is used if analysis or visualization at other
levels (e.g. gene, protein, pathway) is also desired.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">as.seqData</span><span class="op">(</span></span>
<span>  <span class="va">e_data</span>,</span>
<span>  <span class="va">f_data</span>,</span>
<span>  e_meta <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">edata_cname</span>,</span>
<span>  <span class="va">fdata_cname</span>,</span>
<span>  emeta_cname <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  techrep_cname <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>e_data</dt>
<dd><p>a \(p \times n + 1\) data frame of expression data, where
\(p\) is the number of RNA transcripts observed and \(n\) is the number
of samples (an additional transcript identifier/name column should also be
present somewhere in the data frame). Each row corresponds to data for one
transcript. One column specifying a unique identifier for each transcript
(row) must be present. All counts are required to be raw for processing.</p></dd>


<dt>f_data</dt>
<dd><p>a data frame with \(n\) rows. Each row corresponds to a
sample with one column giving the unique sample identifiers found in e_data
column names and other columns providing qualitative and/or quantitative
traits of each sample. For library size normalization, this can be provided as part of f_data
or calculated from columns in e_data.</p></dd>


<dt>e_meta</dt>
<dd><p>an optional data frame with at least \(p\) rows. Each row
corresponds to a transcript with one column giving transcript names (must be
named the same as the column in <code>e_data</code>) and other columns giving biomolecule
meta information (e.g. mappings of transcripts to genes or proteins). Can be the same as edata_cname, if desired.</p></dd>


<dt>edata_cname</dt>
<dd><p>character string specifying the name of the column
containing the transcript identifiers in <code>e_data</code> and <code>e_meta</code>
(if applicable).</p></dd>


<dt>fdata_cname</dt>
<dd><p>character string specifying the name of the column
containing the sample identifiers in <code>f_data</code>.</p></dd>


<dt>emeta_cname</dt>
<dd><p>character string specifying the name of the column
containing the gene identifiers (or other mapping variable) in
<code>e_meta</code> (if applicable). Defaults to NULL. If <code>e_meta</code> is NULL,
then either do not specify <code>emeta_cname</code> or specify it as NULL.</p></dd>


<dt>techrep_cname</dt>
<dd><p>character string specifying the name of the column in
<code>f_data</code> that specifies which samples are technical replicates. This column is used to
collapse the data when <code>combine_techreps</code> is called on this object.
Defaults to NULL (no technical replicates).</p></dd>


<dt>...</dt>
<dd><p>further arguments</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>Object of class seqData</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Objects of class 'seqData' contain some attributes that are
  referenced by downstream functions. These attributes can be changed from
  their default value by manual specification. A list of these attributes as
  well as their default values are as follows:</p><table class="table table"><tr><td>data_scale</td><td>Scale of the data provided in <code>e_data</code>. Only 'counts' is valid for
  'seqData'.</td></tr><tr><td></td><td></td></tr><tr><td>is_normalized</td><td>A logical argument, specifying
  whether the data has been normalized or not. Default value is FALSE.</td></tr><tr><td></td><td></td></tr><tr><td>norm_info</td><td>Default value is an empty list, which will be
  populated with a single named element <code>is_normalized = is_normalized</code>.</td></tr><tr><td></td><td></td></tr><tr><td>data_types</td><td>Character string describing the type of data, most commonly used for lipidomic data (lipidData objects) or NMR data (nmrData objects) but available for other data classes as well. Default value is NULL.</td></tr></table><p>Computed
  values included in the <code>data_info</code> attribute are as follows:</p><table class="table table"><tr><td>num_edata</td><td>The number of unique <code>edata_cname</code>
  entries.</td></tr><tr><td></td><td></td></tr><tr><td>num_zero_obs</td><td>The number of zero-value
  observations.</td></tr><tr><td></td><td></td></tr><tr><td>num_emeta</td><td>The number of unique
  <code>emeta_cname</code> entries.</td></tr><tr><td></td><td></td></tr><tr><td>prop_missing</td><td>The proportion
  of <code>e_data</code> values that are NA.</td></tr><tr><td></td><td></td></tr><tr><td>num_samps</td><td>The number
  of samples that make up the columns of <code>e_data</code>.</td></tr><tr><td></td><td></td></tr><tr><td>meta_info</td><td>A logical argument, specifying whether <code>e_meta</code> is provided.</td></tr><tr><td></td><td></td></tr></table></div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="as.proData.html">as.proData</a></code></p>
<p><code><a href="as.pepData.html">as.pepData</a></code></p>
<p><code><a href="as.lipidData.html">as.lipidData</a></code></p>
<p><code><a href="as.metabData.html">as.metabData</a></code></p>
<p><code><a href="as.nmrData.html">as.nmrData</a></code></p></div>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Rachel Richardson, Kelly Stratton, Lisa Bramer</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pmartR/pmartRdata" class="external-link">pmartRdata</a></span><span class="op">)</span></span>
<span><span class="va">myseq</span> <span class="op">&lt;-</span> <span class="fu">as.seqData</span><span class="op">(</span></span>
<span>  e_data <span class="op">=</span> <span class="va">rnaseq_edata</span>,</span>
<span>  e_meta <span class="op">=</span> <span class="va">rnaseq_emeta</span>,</span>
<span>  f_data <span class="op">=</span> <span class="va">rnaseq_fdata</span>,</span>
<span>  edata_cname <span class="op">=</span> <span class="st">"Transcript"</span>,</span>
<span>  fdata_cname <span class="op">=</span> <span class="st">"SampleName"</span>,</span>
<span>  emeta_cname <span class="op">=</span> <span class="st">"Transcript"</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Lisa Bramer, Kelly Stratton, Daniel Claborne.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.</p>
</div>

      </footer></div>

  


  

  </body></html>

