<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Test for a qualitative and quantitative difference between groups using IMD
and ANOVA, respectively — imd_anova • pmartR</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Test for a qualitative and quantitative difference between groups using IMD
and ANOVA, respectively — imd_anova"><meta property="og:description" content="This is the IMD-ANOVA test defined in Webb-Robertson et al. (2010)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">pmartR</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.4.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Convenience_Functions.html">Convenience Functions</a>
    </li>
    <li>
      <a href="../articles/Data_Normalization.html">Data Normalization</a>
    </li>
    <li>
      <a href="../articles/Data_Object_Creation.html">Data Object Creation</a>
    </li>
    <li>
      <a href="../articles/Filter_Functionality.html">Filter Functions</a>
    </li>
    <li>
      <a href="../articles/Graphing_with_pmartR.html">Graphing with pmartR</a>
    </li>
    <li>
      <a href="../articles/RNAseq_Processing_Workflow.html">RNAseq Data Processing</a>
    </li>
    <li>
      <a href="../articles/Statistical_Analysis_with_pmartR.html">Statistical Analysis with the `pmartR` Package</a>
    </li>
    <li>
      <a href="../articles/Trelliscope_Vignette.html">Using Trelliscope Functions with pmartR</a>
    </li>
    <li>
      <a href="../articles/Typical_Processing_Workflow.html">Typical Processing Workflow</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/pmartR/pmartR/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Test for a qualitative and quantitative difference between groups using IMD
and ANOVA, respectively</h1>
    <small class="dont-index">Source: <a href="https://github.com/pmartR/pmartR/blob/HEAD/R/imd_anova.R" class="external-link"><code>R/imd_anova.R</code></a></small>
    <div class="hidden name"><code>imd_anova.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>This is the IMD-ANOVA test defined in Webb-Robertson et al. (2010).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">imd_anova</span><span class="op">(</span></span>
<span>  <span class="va">omicsData</span>,</span>
<span>  comparisons <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  <span class="va">test_method</span>,</span>
<span>  pval_adjust_a_multcomp <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  pval_adjust_g_multcomp <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  pval_adjust_a_fdr <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  pval_adjust_g_fdr <span class="op">=</span> <span class="st">"none"</span>,</span>
<span>  pval_thresh <span class="op">=</span> <span class="fl">0.05</span>,</span>
<span>  equal_var <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  parallel <span class="op">=</span> <span class="cn">TRUE</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>omicsData</dt>
<dd><p>pmartR data object of any class, which has a `group_df`
attribute created by the `group_designation()` function</p></dd>


<dt>comparisons</dt>
<dd><p>data frame with columns for "Control" and "Test"
containing the different comparisons of interest. Comparisons will be made
between the Test and the corresponding Control. If left NULL, then all
pairwise comparisons are executed.</p></dd>


<dt>test_method</dt>
<dd><p>character string specifying the filter method to use:
"combined", "gtest", or "anova". Specifying "combined" implements both the gtest and
anova filters.</p></dd>


<dt>pval_adjust_a_multcomp</dt>
<dd><p>character string specifying the type of multiple
comparison adjustment to implement for ANOVA tests. Valid options include:
"bonferroni", "holm", "tukey", and "dunnett". The default is "none" which
corresponds to no p-value adjustment.</p></dd>


<dt>pval_adjust_g_multcomp</dt>
<dd><p>character string specifying the type of multiple
comparison adjustment to implement for G-test tests. Valid options include:
"bonferroni" and "holm". The default is "none" which corresponds to no
p-value adjustment.</p></dd>


<dt>pval_adjust_a_fdr</dt>
<dd><p>character string specifying the type of FDR
adjustment to implement for ANOVA tests. Valid options include:
"bonferroni", "BH", "BY", and "fdr". The default is "none" which corresponds
to no p-value adjustment.</p></dd>


<dt>pval_adjust_g_fdr</dt>
<dd><p>character string specifying the type of FDR
adjustment to implement for G-test tests. Valid options include:
"bonferroni", "BH", "BY", and "fdr". The default is "none" which corresponds
to no p-value adjustment.</p></dd>


<dt>pval_thresh</dt>
<dd><p>numeric p-value threshold, below or equal to which
biomolecules are considered differentially expressed. Defaults to 0.05</p></dd>


<dt>equal_var</dt>
<dd><p>logical; should the variance across groups be assumed equal?</p></dd>


<dt>parallel</dt>
<dd><p>logical value indicating whether or not to use a
"doParallel" loop when running the G-Test with covariates. Defaults to
TRUE.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    

<p>An object of class 'statRes', which is a data frame containing
  columns (when relevant based on the test(s) performed) for: e_data cname,
  group counts, group means, ANOVA p-values, IMD p-values, fold change
  estimates on the same scale as the data (e.g. log2, log10, etc.), and fold
  change significance flags (0 = not significant; +1 = significant and
  positive fold change (ANOVA) or more observations in test group relative to
  reference group (IMD); -1 = significant and negative fold change (ANOVA) or
  fewer observations in test group relative to reference group (IMD))</p>
    </div>
    <div id="references">
    <h2>References</h2>
    <p>Webb-Robertson, Bobbie-Jo M., et al. "Combined statistical
analyses of peptide intensities and peptide occurrences improves
identification of significant peptides from MS-based proteomics data."
Journal of proteome research 9.11 (2010): 5748-5756.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Bryan Stanfill, Kelly Stratton</p>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/pmartR/pmartRdata" class="external-link">pmartRdata</a></span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="co"># Transform the data</span></span></span>
<span class="r-in"><span><span class="va">mymetab</span> <span class="op">&lt;-</span> <span class="fu"><a href="edata_transform.html">edata_transform</a></span><span class="op">(</span>omicsData <span class="op">=</span> <span class="va">metab_object</span>, data_scale <span class="op">=</span> <span class="st">"log2"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Group the data by condition</span></span></span>
<span class="r-in"><span><span class="va">mymetab</span> <span class="op">&lt;-</span> <span class="fu"><a href="group_designation.html">group_designation</a></span><span class="op">(</span>omicsData <span class="op">=</span> <span class="va">mymetab</span>, main_effects <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phenotype"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Apply the IMD ANOVA filter</span></span></span>
<span class="r-in"><span><span class="va">imdanova_Filt</span> <span class="op">&lt;-</span> <span class="fu"><a href="imdanova_filter.html">imdanova_filter</a></span><span class="op">(</span>omicsData <span class="op">=</span> <span class="va">mymetab</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">mymetab</span> <span class="op">&lt;-</span> <span class="fu"><a href="applyFilt.html">applyFilt</a></span><span class="op">(</span>filter_object <span class="op">=</span> <span class="va">imdanova_Filt</span>, omicsData <span class="op">=</span> <span class="va">mymetab</span>, min_nonmiss_anova <span class="op">=</span> <span class="fl">2</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> You have specified remove_singleton_groups = TRUE, but there are no singleton groups to remove. Proceeding with application of the IMD-ANOVA filter.</span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> No biomolecules were filtered with the values specified for the min_nonmiss_anova and/or min_nonmiss_gtest arguments.</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Implement IMD ANOVA and compute all pairwise comparisons </span></span></span>
<span class="r-in"><span><span class="co"># (i.e. leave the comparisons argument NULL), with FDR adjustment</span></span></span>
<span class="r-in"><span><span class="va">anova_res</span> <span class="op">&lt;-</span> <span class="fu">imd_anova</span><span class="op">(</span>omicsData <span class="op">=</span> <span class="va">mymetab</span>, test_method <span class="op">=</span> <span class="st">"anova"</span>,</span></span>
<span class="r-in"><span>                       pval_adjust_a_multcomp <span class="op">=</span> <span class="st">"Holm"</span>, pval_adjust_a_fdr <span class="op">=</span> <span class="st">"BY"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(Metabolite)`</span>
<span class="r-in"><span><span class="va">imd_res</span> <span class="op">&lt;-</span> <span class="fu">imd_anova</span><span class="op">(</span>omicsData <span class="op">=</span> <span class="va">mymetab</span>, test_method <span class="op">=</span> <span class="st">"gtest"</span>,</span></span>
<span class="r-in"><span>                     pval_adjust_g_multcomp <span class="op">=</span> <span class="st">"bon"</span>, pval_adjust_g_fdr <span class="op">=</span> <span class="st">"BY"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(Metabolite)`</span>
<span class="r-in"><span><span class="va">imd_anova_res</span> <span class="op">&lt;-</span> <span class="fu">imd_anova</span><span class="op">(</span>omicsData <span class="op">=</span> <span class="va">mymetab</span>, test_method <span class="op">=</span> <span class="st">"combined"</span>,</span></span>
<span class="r-in"><span>                           pval_adjust_a_fdr <span class="op">=</span> <span class="st">"BY"</span>, pval_adjust_g_fdr <span class="op">=</span> <span class="st">"BY"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(Metabolite)`</span>
<span class="r-in"><span><span class="va">imd_anova_res</span> <span class="op">&lt;-</span> <span class="fu">imd_anova</span><span class="op">(</span>omicsData <span class="op">=</span> <span class="va">mymetab</span>, test_method <span class="op">=</span> <span class="st">"combined"</span>,</span></span>
<span class="r-in"><span>                           pval_adjust_a_multcomp <span class="op">=</span> <span class="st">"bon"</span>, pval_adjust_g_multcomp <span class="op">=</span> <span class="st">"bon"</span>,</span></span>
<span class="r-in"><span>                           pval_adjust_a_fdr <span class="op">=</span> <span class="st">"BY"</span>, pval_adjust_g_fdr <span class="op">=</span> <span class="st">"BY"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(Metabolite)`</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Two main effects and a covariate</span></span></span>
<span class="r-in"><span><span class="va">mymetab</span> <span class="op">&lt;-</span> <span class="fu"><a href="group_designation.html">group_designation</a></span><span class="op">(</span>omicsData <span class="op">=</span> <span class="va">mymetab</span>, main_effects <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phenotype"</span>, <span class="st">"SecondPhenotype"</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>                             covariates <span class="op">=</span> <span class="st">"Characteristic"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> See the `group_DF` attribute of your omicsData object or call summary() on your omicsData object to see the combined group levels.</span>
<span class="r-in"><span><span class="va">imd_anova_res</span> <span class="op">&lt;-</span> <span class="fu">imd_anova</span><span class="op">(</span>omicsData <span class="op">=</span> <span class="va">mymetab</span>, test_method <span class="op">=</span> <span class="st">'comb'</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(Metabolite)`</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Same but with custom comparisons</span></span></span>
<span class="r-in"><span><span class="va">comp_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>Control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phenotype1"</span>, <span class="st">"A"</span><span class="op">)</span>, Test <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Phenotype2"</span>, <span class="st">"B"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">custom_comps_res</span> <span class="op">&lt;-</span> <span class="fu">imd_anova</span><span class="op">(</span>omicsData <span class="op">=</span> <span class="va">mymetab</span>, comparisons <span class="op">=</span> <span class="va">comp_df</span>, test_method <span class="op">=</span> <span class="st">"combined"</span><span class="op">)</span></span></span>
<span class="r-msg co"><span class="r-pr">#&gt;</span> Joining with `by = join_by(Metabolite)`</span>
<span class="r-in"><span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Lisa Bramer, Kelly Stratton.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

