<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />

<meta name="author" content="Bryan Stanfill, Kelly Stratton, Lisa Bramer" />

<meta name="date" content="2023-04-19" />

<title>Statistical Analysis with the pmartR Package</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>
<script type="text/x-mathjax-config">
MathJax.Hub.Config({
  TeX: { equationNumbers: { autoNumber: "AMS" } }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>



<style type="text/css">
code {
white-space: pre;
}
.sourceCode {
overflow: visible;
}
</style>
<style type="text/css" data-origin="pandoc">
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
{ counter-reset: source-line 0; }
pre.numberSource code > span
{ position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
{ content: counter(source-line);
position: relative; left: -1em; text-align: right; vertical-align: baseline;
border: none; display: inline-block;
-webkit-touch-callout: none; -webkit-user-select: none;
-khtml-user-select: none; -moz-user-select: none;
-ms-user-select: none; user-select: none;
padding: 0 4px; width: 4em;
color: #aaaaaa;
}
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa; padding-left: 4px; }
div.sourceCode
{ }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } 
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.at { color: #7d9029; } 
code span.bn { color: #40a070; } 
code span.bu { color: #008000; } 
code span.cf { color: #007020; font-weight: bold; } 
code span.ch { color: #4070a0; } 
code span.cn { color: #880000; } 
code span.co { color: #60a0b0; font-style: italic; } 
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.do { color: #ba2121; font-style: italic; } 
code span.dt { color: #902000; } 
code span.dv { color: #40a070; } 
code span.er { color: #ff0000; font-weight: bold; } 
code span.ex { } 
code span.fl { color: #40a070; } 
code span.fu { color: #06287e; } 
code span.im { color: #008000; font-weight: bold; } 
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } 
code span.kw { color: #007020; font-weight: bold; } 
code span.op { color: #666666; } 
code span.ot { color: #007020; } 
code span.pp { color: #bc7a00; } 
code span.sc { color: #4070a0; } 
code span.ss { color: #bb6688; } 
code span.st { color: #4070a0; } 
code span.va { color: #19177c; } 
code span.vs { color: #4070a0; } 
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } 
</style>
<script>
// apply pandoc div.sourceCode style to pre.sourceCode instead
(function() {
  var sheets = document.styleSheets;
  for (var i = 0; i < sheets.length; i++) {
    if (sheets[i].ownerNode.dataset["origin"] !== "pandoc") continue;
    try { var rules = sheets[i].cssRules; } catch (e) { continue; }
    var j = 0;
    while (j < rules.length) {
      var rule = rules[j];
      // check if there is a div.sourceCode rule
      if (rule.type !== rule.STYLE_RULE || rule.selectorText !== "div.sourceCode") {
        j++;
        continue;
      }
      var style = rule.style.cssText;
      // check if color or background-color is set
      if (rule.style.color === '' && rule.style.backgroundColor === '') {
        j++;
        continue;
      }
      // replace div.sourceCode by a pre.sourceCode rule
      sheets[i].deleteRule(j);
      sheets[i].insertRule('pre.sourceCode{' + style + '}', j);
    }
  }
})();
</script>



<style type="text/css">

div.csl-bib-body { }
div.csl-entry {
clear: both;
}
.hanging div.csl-entry {
margin-left:2em;
text-indent:-2em;
}
div.csl-left-margin {
min-width:2em;
float:left;
}
div.csl-right-inline {
margin-left:2em;
padding-left:1em;
}
div.csl-indent {
margin-left: 2em;
}
</style>

<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">Statistical Analysis with the
<code>pmartR</code> Package</h1>
<h4 class="author">Bryan Stanfill, Kelly Stratton, Lisa Bramer</h4>
<h4 class="date">2023-04-19</h4>



<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pmartR)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(pmartRdata)</span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>pep_object <span class="ot">&lt;-</span> <span class="fu">edata_transform</span>(</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> pep_object,</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">data_scale =</span> <span class="st">&quot;log2&quot;</span></span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<div id="introduction" class="section level2">
<h2>Introduction</h2>
<p>This vignette describes the statistical analysis capabilities for
<code>omicsData</code> objects other than <code>seqData</code>. For
details about the statistical methods for <code>seqData</code> objects,
see <code>?diffexp_seq</code>.</p>
<p>The <code>pmartR</code> package includes functions to assess
qualitative and quantitative differences in abundance data using the
IMD-ANOVA method described in <span class="citation">Webb-Robertson et
al. (2010)</span>. Our IMD-ANOVA functionality can handle omics datasets
with various experimental designs, including up to two grouping factors
with or without additional covariate information. The data can be paired
or not.</p>
<p>The functionality of the <code>pmartR</code> package will be
illustrated with peptide data available from the <code>pmartRdata</code>
package called <code>pep_object</code>. In a nutshell, the
<code>pep_object</code> object is a list of length three of the class
<code>pepData</code>. The <code>e_data</code> element contains all of
the peptide abundance data, the <code>f_data</code> object contains
sample information, and the <code>e_meta</code> object contains meta
information related to the peptides in the dataset. See
<code>?pmartRdata</code> for more on this data set.</p>
</div>
<div id="imd-anova-test-for-differential-abundance" class="section level2">
<h2>IMD-ANOVA Test for Differential Abundance</h2>
<p>The <code>imd_anova()</code> function is used to implement the IMD
test, ANOVA test, or combined IMD-ANOVA test. Using the combined test
simply means that both ANOVA and IMD are used, and results reported for
both tests.</p>
<p>Below, descriptions of ANOVA and IMD are provided, followed by a
description of the output from the <code>imd_anova()</code>
function.</p>
</div>
<div id="analysis-of-variance" class="section level2">
<h2>Analysis of Variance</h2>
<p>Quantitative tests for differential abundance of pan-omic data are
accomplished using the function <code>test_method = &quot;anova&quot;</code> or
<code>test_method = &quot;combined&quot;</code> arguments, which implement an
analysis of variance (ANOVA) for each biomolecule. <a href="https://CRAN.R-project.org/package=Rcpp"><code>Rcpp</code></a> is
used to speed up computation (<span class="citation">Eddelbuettel
(2013)</span>) by default, and can be turned off with the
<code>parallel = FALSE</code> argument. Different ANOVA implementations
are applied depending on the complexity of the supplied data. For
example, if only two groups are supplied then ANOVA reduces to a two
sample <span class="math inline">\(t\)</span>-test that assumes equal
variance for the two groups. Alternatively, if two groups are present
and a covariate correction is required, then the effect of the
covariates is removed using a reduced maximum likelihood approach. After
the covariate correction is applied, a two-factor ANOVA is used to
detect difference between all combinations of groups or between main
effects as appropriate. Details on generalizations to these two
approaches are given in the next two subsections.</p>
<div id="one-group" class="section level3">
<h3>One Group</h3>
<p>If there is only one grouping factor and that factor has more than
two levels, then a one-way ANOVA tests for quantitative differences
between the groups. Let <span class="math inline">\(y_{ik}\)</span>
represent the biomolecule abundance for observation <span class="math inline">\(k=1,\dots,n_i\)</span> in group <span class="math inline">\(i=1,\dots,m\)</span> on the log base 2 scale then
we assume <span class="math display">\[\begin{equation}
y_{ik}=\mu_i+e_{ik},
\label{eq:one_group}
\end{equation}\]</span> where <span class="math inline">\(\mu_i\)</span>
represents the mean abundance of group <span class="math inline">\(i\)</span> (on the log-scale) and <span class="math inline">\(e_{ik}\)</span> are the error terms that are
assumed to be independent and identically distributed (iid) according to
the normal distribution with mean <span class="math inline">\(0\)</span>
and variance <span class="math inline">\(\sigma^2\)</span>.</p>
<p>The code below will group the peptide data by Phenotype, which is a
categorical variable with three levels: “Phenotype1”, “Phenotype2”, and
“Phenotype3”. Then the <code>anova_test</code> function fits the model
described above and tests for differential abundance between each pair
of phenotypes for every peptide.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a>pep_object <span class="ot">&lt;-</span> <span class="fu">group_designation</span>(</span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> pep_object,</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">main_effects =</span> <span class="fu">c</span>(<span class="st">&quot;Phenotype&quot;</span>)</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>myfilt <span class="ot">&lt;-</span> <span class="fu">imdanova_filter</span>(<span class="at">omicsData =</span> pep_object)</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>pep_object <span class="ot">&lt;-</span> <span class="fu">applyFilt</span>(</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">filter_object =</span> myfilt,</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> pep_object,</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  <span class="at">min_nonmiss_anova =</span> <span class="dv">2</span>,</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  <span class="at">min_nonmiss_gtest =</span> <span class="dv">3</span></span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>all_pairwise_results <span class="ot">&lt;-</span> <span class="fu">imd_anova</span>(</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> pep_object,</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">test_method =</span> <span class="st">&quot;anova&quot;</span></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>The function <code>imd_anova</code> function returns a data frame
containing:</p>
<ol style="list-style-type: decimal">
<li><p>The number of observed values in each group in the columns called
<code>Count_i</code> where <code>i</code> is replaced by the group name
supplied as an attribute of the <code>omicsData</code>
argument,</p></li>
<li><p>The best linear unbiased estimators of <span class="math inline">\(\mu_i\)</span> in the columns called
<code>Mean_i</code> where <code>i</code> is replaced by the group name
supplied as an attribute of the <code>omicsData</code>
argument,</p></li>
<li><p>The <span class="math inline">\(p\)</span>-values for each
comparison in the columns called <code>P_value_A_</code> (“A” for
ANOVA)</p></li>
<li><p>The (log2) fold change for each comparison,</p></li>
<li><p>An indicator for significance for each comparison in the columns
called <code>Flag_A_</code> (“A” for ANOVA) that is a <span class="math inline">\(\pm1\)</span> or 0 depending on whether that fold
change was or was not statistically significant from zero, respectively,
according to the <span class="math inline">\(p\)</span>-values and the
specified <span class="math inline">\(p\)</span>-value threshold set by
the <code>pval_thresh</code> argument to <code>imd_anova</code>. A value
of <span class="math inline">\(-1\)</span> implies the “control” or
“reference” group is significantly under expressed compared to the test
group, and the converse is true for <span class="math inline">\(+1\)</span>.</p></li>
</ol>
<p>The <span class="math inline">\(p\)</span>-values associated with
each group comparison can be adjusted for multiple comparisons using the
<code>pval_adjust_a_multcomp</code> argument to <code>imd_anova</code>.
A detailed look at the available <span class="math inline">\(p\)</span>-value adjustment methods and when they
should be applied are discussed later in this vignette, but the use of
the Tukey multiple comparison method is illustrated in the
following.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a>all_pairwise_results_adjusted <span class="ot">&lt;-</span> <span class="fu">imd_anova</span>(</span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> pep_object,</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">test_method =</span> <span class="st">&quot;anova&quot;</span>,</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">pval_adjust_a_multcomp =</span> <span class="st">&quot;Tukey&quot;</span></span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>By default, all pairwise comparisons are made. Custom comparisons can
by defined by passing a <code>data.frame</code> to the
<code>comparisons</code> argument with column names “Test” and
“Control”. Now, group “Phenotype1” is treated as the control group and
the remaining groups are compared to it.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a>one_vs_all <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(</span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">Control =</span> <span class="fu">rep</span>(<span class="st">&quot;Phenotype1&quot;</span>, <span class="dv">2</span>),</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">Test =</span> <span class="fu">c</span>(<span class="st">&quot;Phenotype2&quot;</span>, <span class="st">&quot;Phenotype3&quot;</span>)</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>one_vs_all_results <span class="ot">&lt;-</span> <span class="fu">imd_anova</span>(</span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> pep_object,</span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">test_method =</span> <span class="st">&quot;anova&quot;</span>,</span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">comparisons =</span> one_vs_all</span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>If the grouping variable has only two levels, <span class="math inline">\(m=2\)</span>, then a Welch’s <span class="math inline">\(t\)</span>-test is performed instead of an ANOVA.
That is, the model defined in Equation (\ref{eq:one_group}) is still
fit, but the two groups of the grouping factor are allowed to have
different variances. In particular, the error terms <span class="math inline">\(e_{ik}\)</span> are assumed to be independently
distributed according to the normal distribution with mean <span class="math inline">\(0\)</span> and variance <span class="math inline">\(\sigma^2_i\)</span>. To determine if the
biomolecules are differentially expressed, <span class="math inline">\(p\)</span>-values are derived from a <span class="math inline">\(t\)</span>-distribution with degrees of freedom
approximated by the Satterthwaite equation. The value for
<code>Variance</code>, which is not a pooled variance estimate, scaled
by the Satterthwaite approximate degrees of freedom. To illustrate this
the data are grouped by “SecondPhenotype”, which has two values, then
fit with ANOVA.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a>pep_object <span class="ot">&lt;-</span> <span class="fu">group_designation</span>(</span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> pep_object,</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">main_effects =</span> <span class="fu">c</span>(<span class="st">&quot;SecondPhenotype&quot;</span>)</span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>all_pairwise_results <span class="ot">&lt;-</span> <span class="fu">imd_anova</span>(</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> pep_object,</span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">test_method =</span> <span class="st">&quot;anova&quot;</span></span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
<p>When this model is fit, it is imperative the researcher check the
assumptions of the model. In particular for the <span class="math inline">\(m&gt;2\)</span> case, the residuals <span class="math inline">\(\hat e_{ik}=y_{ik}-\hat \mu_i\)</span> are assumed
to be independent, normally distributed and have a common variance
across groups. The appropriateness of this assumption can be assessed
using quantile-quantile plots and comparisons of each group’s
variance.</p>
</div>
<div id="two-groups" class="section level3">
<h3>Two Groups</h3>
<p>Next we consider the case where there are two grouping factors again
in the context of biomolecule abundance. Let <span class="math inline">\(y_{ijk}\)</span> represent the biomolecule
abundance for observation <span class="math inline">\(k=1,\dots,n_{ij}\)</span> in group <span class="math inline">\(i=1,\dots,m\)</span> and group <span class="math inline">\(j=1,\dots,p\)</span> on the log base 2 scale then
we assume <span class="math display">\[\begin{equation}
y_{ijk}=\mu_i+\alpha_j+\gamma_{ij}+e_{ijk},
\label{eq:full_model}
\end{equation}\]</span> where <span class="math inline">\(\mu_i+\alpha_j+\gamma_{ij}\)</span> represents the
mean abundance for an observation in <span class="math inline">\(i\)</span>th first group and <span class="math inline">\(j\)</span>th second group which have marginal
means <span class="math inline">\(\mu_i\)</span> and <span class="math inline">\(\alpha_j\)</span>, respectively. Finally, <span class="math inline">\(\gamma_{ij}\)</span> is the interaction effect of
group <span class="math inline">\(ij\)</span>, and <span class="math inline">\(e_{ijk}\)</span> are the error terms that are
assumed to be iid according to the normal distribution with mean <span class="math inline">\(0\)</span> and variance <span class="math inline">\(\sigma^2\)</span>.</p>
<p>For each biomolecule, the <code>test_method = &quot;anova&quot;</code> argument
automatically tests if the <span class="math inline">\(\gamma_{ij}\)</span> effect is statistically
significant from zero using a full and reduced model <span class="math inline">\(F\)</span>-test. That is, in addition to the model
above, the reduced model <span class="math display">\[\begin{equation}
y_{ijk}=\mu_i+\alpha_j+e_{ijk}
\label{eq:red_model}
\end{equation}\]</span> is also fit and the test statistic <span class="math display">\[F^*=\frac{MSE_F}{(SSE_R-SSE_F)/(df_R-df_F)}\]</span>
is computed where <span class="math inline">\(MSE_\cdot\)</span> and
<span class="math inline">\(df_\cdot\)</span> are the mean square error
and degrees of freedom for the full (F; \ref{eq:full_model}) or reduced
(R; \ref{eq:red_model}) model, respectively. A <span class="math inline">\(p\)</span>-value is computed based on <span class="math inline">\(F^*\)</span> as <span class="math inline">\(p=P(F^*&gt;F_{a,b})\)</span> where <span class="math inline">\(F_{a,b}\)</span> is compared to an <span class="math inline">\(F\)</span> distribution with degrees of freedom
<span class="math inline">\(a=df_F\)</span> and <span class="math inline">\(b=df_R-df_F\)</span>. If <span class="math inline">\(p&lt;0.05\)</span> then there is significant
evidence to reject the null hypothesis that the reduced model
\eqref{eq:red_model} is adequate and the full model
\eqref{eq:full_model} is used to test for group differences. Conversely,
if <span class="math inline">\(p\geq 0.05\)</span> then there is not
enough evidence to reject the null hypothesis and the reduced model
\eqref{eq:red_model} is used.</p>
<p>Regardless of whether the full or reduced model is used, both assume
that the residuals are independent and normally distributed with a
common variance for all groups. These assumptions must be checked before
the results of the model are used to assess differential abundance.</p>
</div>
<div id="covariate-adjustment" class="section level3">
<h3>Covariate Adjustment</h3>
<p>If, in addition to a single grouping factor, there are covariates to
account for when testing for differential abundance then a linear model
of the form <span class="math display">\[\begin{equation}
y_{ik}=\mu_i+\mathbf{x}_{ik}^\top \beta +u_{ik}
\label{eq:covar_adj}
\end{equation}\]</span> is fit to the transformed abundance data <span class="math inline">\(y_{ik}\)</span> where <span class="math inline">\(\mu_i\)</span> is the marginal mean of group <span class="math inline">\(i\)</span>, <span class="math inline">\(\mathbf{x}_{ik}\)</span> is a vector of covariates
for observation <span class="math inline">\(k\)</span> in group <span class="math inline">\(i\)</span>, <span class="math inline">\(\mathbf{\beta}\)</span> is the corresponding
vector of parameters to be estimated and <span class="math inline">\(u_{ik}\)</span> are the error terms. An analogous
model can be fit when there are two grouping factors of interest. If one
of the covariates is colinear with a grouping variable then that
covariate is removed from the model, that is, priority is given to the
grouping variable over the covariate.</p>
<p>In practice, the covariate adjustment is applied then the one-factor
ANOVA model is fit to the covariate adjusted data. That is, after the
vector of parameters <span class="math inline">\(\mathbf{\beta}\)</span>
in <span class="math inline">\(\eqref{eq:cover_adj}\)</span> is
estimated, the response <span class="math inline">\(y_{ik}\)</span> in
<span class="math inline">\(\eqref{eq:one_group}\)</span> is replaced
with <span class="math inline">\(y_{ik}^*=y_{ik}-\mathbf{x}_{ik}^\top\mathbf{\beta}\)</span>.
Though this results in the marginal group means <span class="math inline">\(\mu_i\)</span> being estimated multiple times, it
makes the workflow smoother.</p>
</div>
<div id="paired-data" class="section level3">
<h3>Paired Data</h3>
<p>The comparison of samples at two points in time from the same
individual (e.g., a before and after state) is common in omics
experiments and referred to as “paired data.” To enable analysis of
paired data in <code>pmartR</code>, we have implemented a method for
specifying pair information and added appropriate code to any
<code>pmartR</code> functions impacted by the use of paired data.</p>
<p>If data are paired, this should be specified using the
<code>group_designation</code> function. Users can also designate which
member of a pair is the “control” or “reference.”</p>
<p>ANOVA tests can be performed on paired data with or without any main
effects or covariates.</p>
<p>The robust Mahalanobis distance (rMd) filter for sample outlier
detection has been updated to remove all parts of a pair when at least
one of the individuals in the pair does not pass the specified p-value
threshold. Filters for biomolecules have also been updated to consider
the pair as opposed to each individual member of the pair.</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add fake pair info to f_data using SecondPhenotype</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a>pep_object<span class="sc">$</span>f_data<span class="sc">$</span>Pairing <span class="ot">&lt;-</span> pep_object<span class="sc">$</span>f_data<span class="sc">$</span>SecondPhenotype</span>
<span id="cb6-3"><a href="#cb6-3" aria-hidden="true" tabindex="-1"></a>pep_object<span class="sc">$</span>f_data<span class="sc">$</span>Pair_ID <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="fu">rep</span>(<span class="dv">1</span><span class="sc">:</span><span class="dv">4</span>, <span class="dv">2</span>), <span class="fu">rep</span>(<span class="dv">5</span><span class="sc">:</span><span class="dv">8</span>, <span class="dv">2</span>), <span class="fu">rep</span>(<span class="dv">9</span><span class="sc">:</span><span class="dv">12</span>, <span class="dv">2</span>))</span>
<span id="cb6-4"><a href="#cb6-4" aria-hidden="true" tabindex="-1"></a>mypaired <span class="ot">&lt;-</span> <span class="fu">group_designation</span>(</span>
<span id="cb6-5"><a href="#cb6-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> pep_object,</span>
<span id="cb6-6"><a href="#cb6-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">pair_group =</span> <span class="st">&quot;Pairing&quot;</span>,</span>
<span id="cb6-7"><a href="#cb6-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">pair_id =</span> <span class="st">&quot;Pair_ID&quot;</span>,</span>
<span id="cb6-8"><a href="#cb6-8" aria-hidden="true" tabindex="-1"></a>  <span class="at">pair_denom =</span> <span class="st">&quot;B&quot;</span></span>
<span id="cb6-9"><a href="#cb6-9" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb6-10"><a href="#cb6-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-11"><a href="#cb6-11" aria-hidden="true" tabindex="-1"></a>myfilt <span class="ot">&lt;-</span> <span class="fu">imdanova_filter</span>(<span class="at">omicsData =</span> mypaired)</span>
<span id="cb6-12"><a href="#cb6-12" aria-hidden="true" tabindex="-1"></a>mypaired <span class="ot">&lt;-</span> <span class="fu">applyFilt</span>(</span>
<span id="cb6-13"><a href="#cb6-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">filter_object =</span> myfilt,</span>
<span id="cb6-14"><a href="#cb6-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> mypaired,</span>
<span id="cb6-15"><a href="#cb6-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">min_nonmiss_anova =</span> <span class="dv">2</span>,</span>
<span id="cb6-16"><a href="#cb6-16" aria-hidden="true" tabindex="-1"></a>  <span class="at">min_nonmiss_gtest =</span> <span class="dv">3</span></span>
<span id="cb6-17"><a href="#cb6-17" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb6-18"><a href="#cb6-18" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb6-19"><a href="#cb6-19" aria-hidden="true" tabindex="-1"></a>paired_stats <span class="ot">&lt;-</span> <span class="fu">imd_anova</span>(</span>
<span id="cb6-20"><a href="#cb6-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> mypaired,</span>
<span id="cb6-21"><a href="#cb6-21" aria-hidden="true" tabindex="-1"></a>  <span class="at">test_method =</span> <span class="st">&quot;anova&quot;</span></span>
<span id="cb6-22"><a href="#cb6-22" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="test-for-independence-of-missing-data" class="section level2">
<h2>Test for Independence of Missing Data</h2>
<p>In the event that there aren’t enough observed values to test for a
quantitative difference in abundance between groups, one could still
test for a <em>qualitative</em> difference in groups using the
independence of missing data (IMD) test. This is often the case for
proteomics data where peptides or proteins could have missing data for
one of several reasons. The idea is to assess if there are more missing
data in one group compared to another. If there are an adequate number
of non-missing data available, then the <span class="math inline">\(chi^2\)</span> test of independence can be used.
This assumption often fails, however, so a modified version of the <span class="math inline">\(chi^2\)</span> test, called the <span class="math inline">\(G\)</span>-test, should be used.</p>
<p>The test statistic associated with the IMD test for each biomolecule
is given by <span class="math display">\[\begin{equation}
G=2\sum_{k=1}^K\left[C_{Ok}\ln\left(\frac{C_{Ok}}{E_{Ok}}\right)+C_{Ak}\ln\left(\frac{C_{Ak}}{E_{Ak}}\right)
\right]
\label{eq:gtest}
\end{equation}\]</span> where <span class="math inline">\(C_{Ok}\)</span> and <span class="math inline">\(E_{Ok}\)</span> are the observed and expected
number of non-missing values for group <span class="math inline">\(k\)</span>, respectively. The quantities <span class="math inline">\(C_{Ak}\)</span> <span class="math inline">\(E_{Ak}\)</span> are similarly defined for the
missing (absent) values. Further, the expected number of non-missing
values is <span class="math inline">\(E_{Ok}=(m_On_k)/N\)</span> where
<span class="math inline">\(n_k\)</span> is the number of samples
associated with group <span class="math inline">\(k\)</span>, <span class="math inline">\(N\)</span> is the total number of samples and
<span class="math inline">\(m_O\)</span> is the number of samples with
this biomolecule missing.</p>
<p>The function <code>test_method = &quot;gtest&quot;</code> or
<code>test_method = &quot;combined&quot;</code> argument specification will
perform the test for missing data. The results object returned is the
same as described above, for ANOVA, except an “_G” is used to denote
columns specific to the g-test.</p>
<ol style="list-style-type: decimal">
<li><p>The number of observed values in each group in the columns called
<code>Count_i</code> where <code>i</code> is replaced by the group name
supplied as an attribute of the <code>omicsData</code>
argument,</p></li>
<li><p>The best linear unbiased estimators of <span class="math inline">\(\mu_i\)</span> in the columns called
<code>Mean_i</code> where <code>i</code> is replaced by the group name
supplied as an attribute of the <code>omicsData</code>
argument,</p></li>
<li><p>The <span class="math inline">\(p\)</span>-values for each
comparison in the columns called <code>P_value_G_</code> (“G”
g-test)</p></li>
<li><p>The (log2) fold change for each comparison,</p></li>
<li><p>An indicator for significance for each comparison in the columns
called <code>Flag_G_</code> (“G” g-test) that is a <span class="math inline">\(\pm1\)</span> or 0 depending on whether the <span class="math inline">\(p\)</span>-values were significant or not. In this
case, a value of <span class="math inline">\(-1\)</span> implies the
“control” or “reference” group had fewer observations than the test
group, and the converse is true for <span class="math inline">\(+1\)</span>.</p></li>
</ol>
<p>The <span class="math inline">\(p\)</span>-values associated with
each group comparison can be adjusted for multiple comparisons using the
<code>pval_adjust_g_multcomp</code> argument to
<code>imd_anova</code>.</p>
<div id="covariate-adjustment-1" class="section level3">
<h3>Covariate Adjustment</h3>
<p>Covariates can also be accounted for in the g-test.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a>mypep_covariates <span class="ot">&lt;-</span> <span class="fu">group_designation</span>(pep_object, <span class="at">main_effects =</span> <span class="st">&quot;Phenotype&quot;</span>, <span class="at">covariates =</span> <span class="st">&quot;Characteristic&quot;</span>)</span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>imd_covariates <span class="ot">&lt;-</span> <span class="fu">imd_anova</span>(</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> mypep_covariates,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">test_method =</span> <span class="st">&quot;gtest&quot;</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div id="paired-data-1" class="section level3">
<h3>Paired Data</h3>
<p>The g-test in the imd-ANOVA test can now be run on paired data and
requires a “main effects” designation, unlike the quantitative test
(ANOVA).</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># add fake pair info to f_data using SecondPhenotype</span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>mypaired <span class="ot">&lt;-</span> <span class="fu">group_designation</span>(</span>
<span id="cb8-3"><a href="#cb8-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> pep_object,</span>
<span id="cb8-4"><a href="#cb8-4" aria-hidden="true" tabindex="-1"></a>  <span class="at">main_effects =</span> <span class="st">&quot;Phenotype&quot;</span>,</span>
<span id="cb8-5"><a href="#cb8-5" aria-hidden="true" tabindex="-1"></a>  <span class="at">pair_group =</span> <span class="st">&quot;Pairing&quot;</span>,</span>
<span id="cb8-6"><a href="#cb8-6" aria-hidden="true" tabindex="-1"></a>  <span class="at">pair_id =</span> <span class="st">&quot;Pair_ID&quot;</span>,</span>
<span id="cb8-7"><a href="#cb8-7" aria-hidden="true" tabindex="-1"></a>  <span class="at">pair_denom =</span> <span class="st">&quot;B&quot;</span></span>
<span id="cb8-8"><a href="#cb8-8" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-9"><a href="#cb8-9" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" aria-hidden="true" tabindex="-1"></a>myfilt <span class="ot">&lt;-</span> <span class="fu">imdanova_filter</span>(<span class="at">omicsData =</span> mypaired)</span>
<span id="cb8-11"><a href="#cb8-11" aria-hidden="true" tabindex="-1"></a>mypaired <span class="ot">&lt;-</span> <span class="fu">applyFilt</span>(</span>
<span id="cb8-12"><a href="#cb8-12" aria-hidden="true" tabindex="-1"></a>  <span class="at">filter_object =</span> myfilt,</span>
<span id="cb8-13"><a href="#cb8-13" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> mypaired,</span>
<span id="cb8-14"><a href="#cb8-14" aria-hidden="true" tabindex="-1"></a>  <span class="at">min_nonmiss_anova =</span> <span class="dv">2</span>,</span>
<span id="cb8-15"><a href="#cb8-15" aria-hidden="true" tabindex="-1"></a>  <span class="at">min_nonmiss_gtest =</span> <span class="dv">3</span></span>
<span id="cb8-16"><a href="#cb8-16" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb8-17"><a href="#cb8-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb8-18"><a href="#cb8-18" aria-hidden="true" tabindex="-1"></a>paired_stats <span class="ot">&lt;-</span> <span class="fu">imd_anova</span>(</span>
<span id="cb8-19"><a href="#cb8-19" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> mypaired,</span>
<span id="cb8-20"><a href="#cb8-20" aria-hidden="true" tabindex="-1"></a>  <span class="at">test_method =</span> <span class="st">&quot;gtest&quot;</span></span>
<span id="cb8-21"><a href="#cb8-21" aria-hidden="true" tabindex="-1"></a>)</span></code></pre></div>
</div>
</div>
<div id="output" class="section level2">
<h2>Output</h2>
<p>The <code>imd_anova()</code> function returns an object of class
<code>statRes</code> that is a data frame with some attributes. The data
frame includes:</p>
<ul>
<li>Biomolecule name</li>
<li>Counts of observed values per main effect group</li>
<li>Means of observed values per main effect group</li>
<li>log2 fold changes for each comparison</li>
<li>p-values for each comparison and test</li>
<li>Flags indicating significance</li>
</ul>
<p>The <code>statRes</code> object also consists of several attributes
which specify the methods used to arrive at the results. This way, the
results object is self-contained and can be passed to other functions
for further analysis with ease. In particular, the attributes are</p>
<ul>
<li><code>group_DF</code> - <code>data.frame</code> defining which
samples are associated with which groups</li>
<li><code>comparisons</code> - <code>data.frame</code> defining the
comparisons that were tested</li>
<li><code>number_significant</code> - <code>data.frame</code> giving the
number of significant biomolecules (up and down) for each
comparison</li>
<li><code>statistical_test</code> - character string giving the
statistical test(s) run</li>
<li><code>adjustment_method_a</code> and
<code>adjustment_method_g</code> - character string giving the
adjustment for multiple comparisons that was used; NA if none was
used</li>
<li><code>pval_thresh</code> - the numeric value that was an input to
<code>anova_test</code></li>
</ul>
<p>Several methods are defined for objects of class <code>statRes</code>
including <code>summary</code>, <code>print</code> and
<code>plot</code>, which are discussed in more depth in the next
section.</p>
</div>
<div id="methods-for-results-objects" class="section level2">
<h2>Methods for Results Objects</h2>
<p>Special functions are available for objects of class
<code>statRes</code> including <code>print</code>, <code>summary</code>
and <code>plot</code>. Due to the typical size the <code>statRes</code>
objects, only the head of each element in a <code>statRes</code> object
is printed along with the summarized attributes; use
<code>print.default</code> to see the full object. The
<code>summary</code> functions prints the type of test that was run, any
adjustments that were made, the <span class="math inline">\(p\)</span>-value threshold used to define
significance and a table that summarizes the number of significant
biomolecules (up or down) for each comparison. See below for an
illustration.</p>
<p>The <code>plot</code> function can be used to produce any subset of
the following four plots specified by the <code>plot_type</code>
argument. If not <code>plot_type</code> is supplied, then all four are
created, otherwise on the subset specified is produced.</p>
<ul>
<li><code>&quot;bar&quot;</code> - shows the number of significant biomolecules
grouped by comparisons</li>
<li><code>&quot;volcano&quot;</code> - a plot of the <span class="math inline">\(p\)</span>-value by the fold-change estimate
differentiated by test and faceted by comparison</li>
<li><code>&quot;heatmap&quot;</code> - illustrates the fold changes for the
statistically significant biomolecules (only available if more than one
comparison is done)</li>
</ul>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" aria-hidden="true" tabindex="-1"></a>stat_results <span class="ot">&lt;-</span> <span class="fu">imd_anova</span>(</span>
<span id="cb9-2"><a href="#cb9-2" aria-hidden="true" tabindex="-1"></a>  <span class="at">omicsData =</span> pep_object,</span>
<span id="cb9-3"><a href="#cb9-3" aria-hidden="true" tabindex="-1"></a>  <span class="at">test_method =</span> <span class="st">&quot;combined&quot;</span></span>
<span id="cb9-4"><a href="#cb9-4" aria-hidden="true" tabindex="-1"></a>)</span>
<span id="cb9-5"><a href="#cb9-5" aria-hidden="true" tabindex="-1"></a><span class="fu">summary</span>(stat_results)</span></code></pre></div>
<pre><code>## Type of test: combined 
## 
## Multiple comparison adjustment ANOVA: 
## 
## Multiple comparison adjustment G-test: 
## 
## p-value threshold: 0.05 
## 
## Number of significant biomolecules by comparison. Columns specify fold change direction and type of test:
## 
##   Comparison Total:Positive Total:Negative Positive:ANOVA Negative:ANOVA
## 1     A_vs_B           2062            287           1038            108
##   Positive:G-test Negative:G-test
## 1            1024             179</code></pre>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="fu">print</span>(stat_results)</span></code></pre></div>
<pre><code>## statRes
##                         Peptide Count_A Count_B   Mean_A   Mean_B
## 1              AAAAAAAAAAGAAGGR       4       6 26.82222 26.19385
## 2                   AAAAAAALQAK       3       3 25.76461 26.51143
## 3     AAAAAEQQQFYLLLGNLLSPDNVVR       8       8 25.02951 24.83093
## 4                  AAAAAGTATSQR      10      11 25.35116 25.37603
## 5                          ----    ----    ----     ----     ----
## 18103            YTIIIPENLKPQMK       5       1 26.16265 26.18524
## 18104                  YTQELTLK       3       2 25.07080 24.42717
## 18105             YTYSEWHSFTQPR       2       2 25.21807 25.38213
## 18106            YVYDSAFHPDTGEK       3       2 25.07202 24.09574</code></pre>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(stat_results)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAkAAAAHgCAMAAAB6sCJ3AAAB7FBMVEUAAAAAADUAADoAAGYAOgAAOjoAOmYAOpAAZgAAZpAAZrYAkAAAtgAA2wAA/wAZGUgZGXEZSJcZcboaGhozMzM1NYQ1hMk6AAA6ADo6AGY6OgA6Ojo6OmY6OpA6ZpA6ZrY6kLY6kNtIGRlIGUhIGXFISJdIl91NTU1NTW5NTY5Nbo5NbqtNjqtNjshmAABmADpmAGZmOgBmOjpmOpBmZgBmZjpmkJBmkLZmkNtmtrZmtttmtv9uTU1uTW5uTY5ubqtuq+RxGUhxGXFxSJdxcXFxuv9/f3+OTU2OTW6OTY6Obk2Obm6OyP+QAACQOgCQOjqQOmaQZgCQZjqQZpCQkDqQkGaQtpCQttuQ27aQ29uQ2/+XSBmXSEiXSHGXupeX3bqX3f+o6+urbk2rbm6rbo6rjk2ryKur5Mir5P+2AAC2ZgC2Zjq2Zma2kDq2kGa2tma2tra225C229u22/+2/7a2/9u2//+6cRm6///Ijk3I///JhDXJ6+vbAADbkDrbkGbbtmbbtpDb25Db27bb2//b/7bb/9vb///dl0jd/7rd/93d///kq27k/8jk/+Tk///rqF7ryYTr68nr6+v/AAD/tmb/unH/yI7/25D/27b/29v/3Zf/5Kv//7b//7r//8j//9v//93//+T////VNH6GAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAe1ElEQVR4nO2djZ8bx1nHFznJubLvgCYgX4tTrpDIqYmSQF+jxjavTuByhkBCe+WlOKWcFQqENNf6StSGll6ghqO1FSvJiVNv/1HmbXdndjU7O9rV6tH69/18fCvt2zP7zNczs6tdKQgBKEGw7AKA1QYCgVJAIFAKCARKAYFAKSAQKAUEAqWAQKAU5QQaBltiOu6081YbnNkvvs/Jc0HQF69GgeDhT/1Ie8eIYkXLv3AYhtOdtcNi+y+wpmMVPa7Bh8VKkOWHr3Rm7W8VKCtQa5dPqxRoEKidJsqI91aBxIwlCJQURPH9TY/D1MO9oR/nilFWoECkukKBpjvxyiPZEn34RsBnqXcaas4Hz3klvgqBZsf1+n+ibxc8/PxhfJwrRkmBWp8TyaxUoLjyYmUGvKe0CsSmW4V3X6FAmbhzCjTuPKK2G2YOkT5lBfpKj2eNCySzLv4O1n78StD6Yjh9JQg+dcgz+zZrpX/tLb7JB2xeS45a2sPgzFtyRx/wQcDjb6nuQdkYKzPptfMF6qvo8W4yRUgWRWuqYvCWJAguxgv43/Qq0zc6UfHTcZPVpjus5B8TWwqzxE70MMmBr/37c7xsURLjA5v8+VfjtKTKK/Oa5DFVpOVRVqBdkayMQH/GPXiep5QvHpz5bNTFjzvJqOXhTqD+p6u5zMWZAomd5nRhbDuxSrKbTBGSRfqa7Th2tCARSFtlYI5Q9LjJakKg9pCvNeBbTXpb2TDywOWIZ0s7uASVllR5I4HiPKaKtDxKCxQO2L+0QMHaW+G7Hf73B4F4f+arfKjY5gf+64f8JW8zkh5gwFsJucKsLkx0DuaQWi5XM/px3Hg32SJEi9SaRjGm34obsVigZJVJjy/4QVxcPa62miol39En+SueG22hHrF9yNod9d9ENK8JKi2p8kalbv2lnJku0vIoL9C4w/+/mAfK61iOhpP3YoYaLLF2Wh8ty3TIOXMIJHLNNtN3kyqCtoi/1YqR1J8pkLFK6/F/+VFy1HpcbTVRSr67cefzF3b5O21h5sBViWKBJj2+Rx5bWxqVN5PHdJGWR3mBeB+eFihWR3sv1lYts8xUbMq4sxXvbpZA+V0Yaz62QlnnyW5SRdAWybdxMaIlWYHiVfjJZjyUScXVVxvIvvNweObbvb40SQuTOvD4ONWLRCD+NlXeTB7TRVoeFQg06Z1501ug6H+WwCWQ+O9rH0Qnde4tEG8lHALxHf3wFeVIJq6+mog5OrM/aIeD9rjTD40wqQNPjnMQN6nJAN4pUKpIy6MCgVhCH+4YLXNGIK0Li3ttzRRXFzbgL/IEkrVidGFmETJdWFwMvQuLDiC1iuDDd+PLPnpcfbWBHFR/iY2eh2vfYm+0hZkDT46TjwGSJXK+Ud4kr3Ee00VaHlUIxE8J2vJqDRsCzhIoeOStaETY+iIf/cUtliR3EP3hG4E8N7Z3YfHQOBlEZ4qgD0qNYgTPh2y4rR9AahXWALLhxgc7KYFU3Hi1aNz1STYAGrU+IY5FC5M6cO042dj4CyzA9N3nkp7dOB+I8xrnMV2k5VGJQKx9bqux5SPPzezCPht1BqopN/sqba4pkD5sTj7KiAbf+nL9zFuc5qaKkIqrrSlf8lfxAWR2JsNkPkLRrkvIwSBfZRiIsbS6wDpjvYxAYfRRhvBjR13bj7fU8tr6RHT8qSItj0oEUuek3+8Ej/949hjobdZjPx5fT9MukCnEVTbxoWlWIPUho1WgVuryoNhNugjJouStvOQoLiQ+Il5FB5BeZfpP/LPO58MwG1dfbcJbEDlgm6q2QQ9jHnjq8Dt8f/+qHX9yKHpe4zymirQ8cDvHCjHvh22LBAKtEBAIlAICgVJAINA4IBAoBQQCpYBAoBQQCJQCAoFSlBLoMVAbVVV41ZQTqKpSABdkUw2BVgOyqYZAqwHZVEOg1YBsqiHQakA21RBoNSCbagi0GpBNNQRaDcimuukC/d/vfj1v8UeffpTxh3WVZn7IprrpAr3/B0/mLf7oN7/L/vxGrmQkIJvqhgv087/629//iXr90RM/Ye+//v6jj/7Sd6PlQiA2c0nFKw7ZVDdcICbNd6IOinvy0RP/+XvfDd/7eLxctEBP/MSyOR3IprrhAr33ZPj+x7U37z1pDorkGOiXIdDcNFugn3/tUa3D+uiJ/+KN0Kcf/cXYIdmFfY3+KJpsqpstkOictD7sb2Rn9X7c5AiBkjXoQjbVzRboPW6G1oc9+qSQJyWQ41SfBGRT3WiBfv7XvH1J/BDn698xzsJwHagkjRaoQZBN9YMgkGxmtKHzzFm0IZvqB0GgJkA21RBoNSCbagi0GpBNNQRaDcimGs+FrQhVVXjVLL0FulvBPohHrCIgBLIBgQoBgWxAoEJAIBsQqBAQyAYEKgQEsgGBCgGBbECgQkAgGxCoEBDIBgQqBASyAYEKAYFsQKBCQCAbEKgQEMgGBCoEBLLRUIGmO4H6Fcd+eFf8lKGaMycQyEZDBRpyY/qTZ3bHm/s/vXjIXss58+4PAtloqECc4daYqXNj93tbyZx59wWBbKyaQBvFCTbOn93YOHv+7LkgOKvmFCQTFgLZaK5AwXok0EP8r5wDgXQgkJ11bss6U+ehdS7PuXNyDgTSgUA5/ohJsL4edWTrhfsvCORDQwU6y39Z/tzG+SDgzQ9/qeZAIA0ItCgyYVdRoOMvd7vbYXhyvXv5KD2RQCAIZBfo5MW98PiFvdOb2+Gdp0NzooBAEMgu0H3uye3tk5cOwuMrB+ZErQKBIFD+GIi1QsdXj7KTUD2VevfBox6BMmFXU6DTm9fC+5eFMuZELUcLtCgyYVdSoJPr19hQ2tICcSAQBMo9C2PnYCHGQCYQyMQukPRHdGPi9EufKCAQBLILdKfL2cZ1IBMIZIIr0Z5AIBMI5AkEMoFAnkAgEwjkCQQygUCeQCATCOQJBDKBQJ5AIBMI5AkEMoFAnkAgEwjkCQQygUCeQCATCOQJBDKBQJ5AIBMI5AkEMoFAnkAgEwjkCQQygUCeQCATCOQJBDKBQJ5AIBMI5AkEMoFAnkAgEwjkCQQygUCeQCATCOQJBDKBQJ5AIBMI5AkEMoFAnkAgEwjkCQQygUCeQCATCOQJBDKBQJ5AIJNyAi3hSy6XTT0CZcI2VKAKCoAWaBaZsBDIBgSCQKWAQBCoFBAIApUCAkGgUkAgCFQKCASBSgGBIFApIBAEKgUEgkClgEAQqBQQCAKVAgJBoFJAIAhUCggEgUoBgSBQKSAQBCoFBIJApYBAEKgUEAgClQICQaBSQCAIVAoIBIFKAYEgUCkgEAQqBQSCQKWAQBCoFBAIApUCAkGgUkAgCFQKCASBSgGBGizQ8ZWDMLzT7XafOghPrncvH4XRRAKBIFCeQPe5OOHtbf769OZ2eOfpaKKAQBAoR6Dbl77JWqDT1/f4m5OXDniDpCZqDQgEgZxdGOuzut3t8PjqUXjy4p6asGWPcZbwHYXLph6BMmFXV6DjF/Z4K3T/sjBHTdRytECLIhN2dQUS3N7OtEAcCASBigqEMZACApk4BeKd1uk3Dk5vXpNnYddwFgaBNApdB7q0F+I6kAICmeBKtCcQyAQCeQKBTCCQJxDIBAJ5AoFMIJAnEMgEAnkCgUwgkCcQyAQCeQKBTCCQJxDIBAJ5AoFMIJAnEMgEAnkCgUwMgSa9/qQXnNkvujEEgkCGQIN2ODyzP2wX3RgCQSBdINYATXfa4ahwEwSBIFBKoElvCwLlAoFMdIGmO1uj1i7vyAoCgSCQMQYad4J2OFg7LLoxBIJAOI33BAKZQCBPIJCJKdAwCPpDdGF5QCAT8zrQ2n/IM/mCQCAIlDmN7+M0PhcIZAKBPIFAJkYXNuRdGL+WWBAIBIHMQfQoYBT2BwJBIJzG+wKBTCCQJxDIJBZo0gsiMIjOAQKZoAXyBAKZQCBPIJCJeR3ItwtbwpdcLpt6BMqEXQmBJJNndotujBZoUWTCrpBA4ajwp6kQCALNEghnYTlAIJMZAhW/JRECQaAZg+gWxkA5QCATnMZ7AoFMIJAnEMjEEGi6sxV63JAIgSBQ5tHm0McgCASBUnck8glO4/OAQCYQyBMIZGLe0srVwS2tuUAgE9zS6gkEMsFpvCcQyAQCeQKBTPBosycQyASPNnsCgUzwZKonEMgEAnkCgUzwaLMnEMgE14E8gUAmOI33hJRA0x35HE0/mjG+IO8GTIYhQ3WLYLSkGMW3gkCeEBMo1V2kBZru8IsyQ2aYl0DiM61R4mUOeLTZk9USSN3fPlw79BFoutPXNs4HLZAnRAXinVlbCsSagtarUiB1fwVnfOFV2deNO2KcO77wWke856u/trkv9hC3W1vxVrvRCvEGJhDIE5oC8au//B+rcH4WPelJFbRmZ9xh7QnrmoRTbJq8l6uL68fRpxCj6FRK21+0QaoImY8ycBaWDzGBAlljostif1iFi5eqosebcX2PO33hw4dcETaN3kerj+StPFELw01oGytEG6SKsKT7gSbPilEaaxHvDkS7OBTlrYVGCaTqSjxNzLW4sCtaEWWO3gJdiE6qRuIMK3ofrT4MglTrMemJkVO0QrIDg+XckSia2Mkzu6xc/9vmNo0vHqqR2+J5gASKWhSealX/kx47PZdrmgJlR8zm/igJNP3Tb7MWiEtzY/duGAl0w+dChTdJm/ezHdXaDYt31glEBeIXbbQuLKpDpcXwTFz/QrVR0gJFq4+050mVJlIgrXdcfBdWPEPBxsb5sxsbZ89vbJwL2J/1IFifO7sF0Nq877Xleeq40xyBjEF0Ox5Ep68DcR94A9RpaS1SNIhma0YWDfg0NSgvIlDpQXTxDGkC8b/nH9rYeOj8vNl1k2rzeOMzvfG55ghkOY1nDPRryvwPm9H6Sq8fv49W53uIW6FIBG2FQgKlOL5yEIYn17uXj9ITy1EVzxATaJ1LI5qdcwwplG92PerEUDaIgi4yYgkKClQVzm/0sa+QJ9D97lMH4enN7fDO06mJopxAG8H6eqAqlbdAZ4vW51wBDYFYd8n0hUChHD3l3kPoWCH9g3PaRxm3L32TtUAnLx3wlsicqPVLCnQ+4KOfs0FwTvwt2gDNL5Bq89b5cOscP9LFKluCTD0trgVSV3zmXcH8ycvUelyV46tH4cmLe+aELXuMs/QvEPTYKm7z1gM1p3gLhO9ItDLjND6BC3T/slDGnKjlJVqgMswXMG7zzkZtD7qwCjBboBkC2VogzkoJtFIRM/W0EgJlLiEeL24MVFF26wkIgezkDKKlQKc3r8nTL32igEA1BLSkmgqp3403Fy7wOlBF2a0nIASykzuIdgGBaghoSbXk3iz8KrEkuYNoFxCohoCWVEtoCaTff1QICFRDQEuqJbQE8v+xlfSM2rNbT0AIZKfae6Jrz249ASGQHQhEMqI71RJqAtV3P1BF2a0nIASys6w7EivKbj0BIZCdau+Jrj279QRcUYHEgzjWU2t+n7znWfcsIBDJiO5USxwC8ZrMEahYLeeDLoxkRHeqJfkCbf5zW4oy6QWyaoNHfntXfOTZ57Pe3Hzz4mE4vbGrls8DBtEkI7pTLXEItD/oC4EGW+GQucQmo9Yu/1Fc2X+NN9++scsfNVDL5wGn8SQjulMtcQk0+Z1D/pcpM3l2nz8ap56+44/iCYeGTJ0ttXwuByAQyYjuVEtcAjE7uEA9+TSz7K/Eoz5nlEDjiz8WPZjPD1UaaN8P1MdHGWQiulMtcQo0vfHa5r5qXVQLxM+Voi5sny2/eDhv68NBC0QyojvVEqdA8lSMD37WDtUYiM/mTy3L0/ghH/Oq5fMAgUhGdKda4hZInFrzh0t3xVnYx1gXxk6VHnmuP91hZ2H78llTtXweNIFYz7jFz/BK/ORl7dmtJ+CKCjSLiq7+JCQCDdYOpztBP3tjqx0IVENAS6olXgIZT79XRSyQuB1RfpVH4d4QAtUQ0JJqiXcLVDn6WVgYGt8t4wYC1RDQkmoJBCqZ3XoC0hVo+UAgkhHdqaYCBCIZ0Z3qnNIUrL1qqPab6mvPbj0BIZAdXEgkGdGd6pzSlKlSbyAQyYjuVOeUpkyVegOBSEZ0pzqnNGWq1BsIRDKiO9U5pSlTpd5AIJIR3anOKU2ZKvXGvB8IX65AJKI71Tml8avEkpQTaJW+ZHOlIhb9ks2ZG0cL5feFpT8aj+8l86trG0kXNsR1IDoR3anOKU20UEjC73dOU+U9HfiCKZIR3anOKU20UHjC77JQT+2MRNsw3vx2RU/0CDCIJhnRneqc0kQL5S2t/OcxxB2r/M7nYbvCJ3oEeC6MZER3qnNKEy2UY6C+uJme9WSqM4vHQKWf6BHgyVSSEd2pzilNtDAa6sinM8QTqa3kmcLyT/QI8Gw8yYjuVOeUJloYCaRaIFG3a4fVPdEjgEAkI7pTnVOaaGF8siXHQPxWZV2g0k/0CNCFkYzoTnVOaaKFsUDqTGugzsL2K3qiR4BBNMmI7lTnlKZw/VUBTuNJRnSnOqc0ZarUGwhEMqI71TmlKVOl3kAgkhHdqc4pTZkq9QYCkYzoTjUVIBDJiO5UUwHXgUhGdKeaChCIZER3qqkw636gwh/OQqAaAlpSTQXcD0QyojvVVMAgmmREd6qpkPujuy4gUA0BLammgvmTl563pkGgGgJaUk0FjIFIRnSnmgr40V2SEd2ppoIxBip+BUgCgWoIaEk1FfCjuyQjulNNBZzGk4zoTjUVIBDJiO5UUwFdGMmI7lRTIdsCzXqY2gIEqiGgJdVUmNGF4Zvqlx/RnWoqzBIIXdjSI7pTTYUZAg3QAi09ojvVVJgxiC7+nBkEqiGgJdVUcJ/G3+l2u08dhCfXu5ePwmgigUA1BLSkmgpugW5v87+nN7fDO09HEwUEqiGgJdVUcD7afPr6Hp+cvHQQHl85UBO1DALVENCSaio4v1yB9Vnd7nZ4fPUoPHlxT03Y/Mc4+JLNxQfM/ZLN5eN8KuP4hT3eCt2/LMxRE7UMLVANAS2ppkKxx3pub2daIA4EqiGgJdVUKPb9QLe3MQZa1iFaUk0F5yCad1qn3zg4vXlNnoVdw1lYvQEtqaZCoetAl/ZCXAda1iFaUk0F3A9EMqI71VTQBBJj6OmOxzcuQqAaAlpSTYVEoHFHDn4Gxe+sh0A1BLSkmgqJQIN2+oUTCFRDQEuqqWD+7LcA9wMtP6I71VSAQCQjulNNhVig6U50/WeIG8qWHtGdaipoXzClGp7EJCcQqIaAllRTQTuNH4hbESe94ufxEKiGgJZUU0G/kCi+HsjnhxMgUA0BLammAq5Ek4zoTjUVIBDJiO5UUwECkYzoTjUVIBDJiO5UUwECkYzoTjUVIBDJiO5UUwECkYzoTjUVIBDJiO5UUwECkYzoTjUVIBDJiO5UUwECkYzoTjUVIBDJiO5UUwECkYzoTjUVIBDJiO5UUwECkYzoTjUVIBDJiO5UUwECkYzoTjUVIBDJiO5UUwECkYzoTjUVIBDJiO5UU6GcQPiOxMUHXKHvSPQGLVANAS2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppgIEIhnRnWoqQCCSEd2ppoKvQCfXu5ePojcQqIaAllRTwVOg05vb4Z2no3cQqIaAllRTwVOgk5cOwuMrB+odBKohoCXVVPAU6PjqUXjy4h579Rjnj0BNNEWg+5cjgTiZo7pXC7UHrD9iJvFNEShpgTgQqIaAllRTodoxUO3ZrScgBLLjfRZ2Le8srPbs1hMQAtmp9jpQ7dmtJyAEslPtlejas1tPQAhkBwKRjOhONRUgEMmI7lRTAQKRjOhONRUgEMmI7lRTAQKRjOhONRUgEMmI7lRTYYkC9W6xP0HQu3fvnSAI5suuV8D5qT2iO9VUWJ5AvYAJ1Hv5XufWvc47917uzZVd34qZk9ojulNNhaUJ9PIt3gJxdV5Wf+fJrnfNzEftEd2ppkK1AvkweXY/HLXDu4N+OO60dufbST3VCYHsUBBotHY43enPtZN6qhMC2VmuQOOLhz+7sTvcCrlK81BPdUIgO8sVaPLM7k8391kLFA625tpJPdUJgewsV6BwFASs7xoEwXwNEARaOssTSHG31Nb1VCcEsgOBPOuz9oASCGSjToF6t8RF76Bz75a4Aj5PffpFnJfMgUIgGzUKJK59Mzq37gXEL11mDhQC2ahPIHnt+969W71773TuiX9z1KefCPOSOVAIZKOcQF6I875w/CuHobh0SfnCQSYsBLJRu0CDX+V/gtbnIFAVPHACTZ79O/luQPnDk0xYCGSjboHGF/9NfHAy3tyfaycQyOSBE2jU5hEH4gr4PEAgkwdJoEoiQiATCOQJBDKBQJ5AIBMI5AkEMoFAnkAgk3IC3X3wqEegTNiGClRBAdACzSITFgLZgEAQqBQNFuhl8cCteOxW3YYEgUwgUB7vMHte7iWP3XZuQaAUEMgpUSd+7PbWnLdASiCQjVUTyAvx7FvnF/hjt+OLh2X2BIFsNFmgyTO7/N61n/HHbud88C0CAtlosEDjC6zpGW6Fd0dtdTPb/EAgG80VSN5xxFqgu6z1KdmDQSArzRVowE/dt9Rjt3M++x8DgWw0V6BKA0IgGxCoEBDIBgQqBASyAYEKAYFsQKBCQCAbEKgQEMgGBCoEBLIBgQoBgWxAoEJAIBsQqBAQyAYEKgQEsgGBCtFQgUBtVFXhVVNKoCqoPzO1RyRb+VUAgRoYsE4gUAMD1snSBQKrDQQCpYBAoBQQCJQCAoFSPEgCHV85cMxYRKxZrxoEBFp0LAhULcdf7na3w+Orf88n4cn17lMHx1ePwtPX9+SShcDCfOZP/oHHEgHD8H43eb2ggGr/8gjlQUOg8py8uMcTefzla+H9y0fhbT75n9fZvKv/LZcsJCoPc2mP7529uvN0ePLSAZ8suAUSB3fEXkUHDYGqgVUfzyXPLKtIlt07rFavqSULCiiaOFmV7B2fhIvuwtTBqSDqoBtH/QLd5r2HEohPeM1e5Y2QXLKIkFEfeYV3KN3upT3ekV5aXIMXhvrB8SDxQTeO2ruw69tRcpP/pKevf/PqkVqykKBJC5Q0cbJ3WUQ4jtECaQfdOOoWSJjzwp4SSA0Twjvda9GShUQ1xkAsII+5cIGSMZB20I2j9i7sTrf7mT/ejgRSJ0JCHLlkIUFZmN96fe/0pjgpuhT3lnzGQuKFoYqlgiQH3TgemOtAjaw9AjwQAp3eFCNnsAAeCIHA4oBAoBQQCJQCAoFSLFmgYRAELf493B+8lcxkr8U35M5+q4hm6RtqW6RXq4d6o5FguQINz/DfUQ76RuZTtWCpFDV71tKcnYGqWapA0x3x29uDtUMItKosWSD1/f/jDv9GZf432BKveb2Lt/347XRHfOGyWkuZITfki3hbpm0RhRhfeDUI1g7DAd90KL+sedILWq9t7o83/4JtpXYrdsf+ROur8smQ8TqvdVQB+/GO4/Lwvb0ZFzu10Zw/Ur8CLLcLG0VVzetv0uuLTk3Iw6uSV2mnr95Od9piDX0ttSFfFA5VM6a2iBh31g758hETTDV4k94W+8f20GmHYlv+Lxaow52SokUh43XYvtiojf/hW8sV4/J02qFebHOjM+V+6IAwyz4L46No1QB8yP/fqyrgk02ZdN0nhr6WWsrl4BWpbxHBqlLM4PWslovVRZX31ZuR8lEI1M90j/E6YplaIZrG5Ynmq0KYGzW3I122QCGv+mgMNBLnZMoY9RsBkUCj+D9xspZayiXUflXAFGhTysW7L9WDDUXHsymdGYk3UcOhpqJVUQ6E2XVSK2rl0Yqd3aiZEBAoSvOk19rVWqBQDFbiHi2qTX2tUAm0Fv+QSbxFvO9IoPHm27IHq1ggozxasSFQDSQSiJ6Ip3zU0gSSHZPZhelrqV2MWlr1qC2MEHKg9Pl05yIE4ttmujC5ZhwytY6xolEerdjZjZrJclugAc9yNBTlKR93WnK4opqcaPQiB9FiNKzWUnUiB7msEtl8fYuIaBDNR1vqB3PiQfSF3WQQzWdOd1qZQbT+LyWQXNEoj1bs7EbNhMCVaHEiNmDVy0YPra+wymevRcOirlKrt9E5tVorqhO+IV8k/sMnW0RXCMTZtvBBNBkCfhr/atTT6VcHvvTMbrJ+OOM03ujCXs2WRy92ZqNmQmEMtBDGX0zP2NRPpUdrs38ArnBVN9gJLxor0Cj9G6XDaAbvdKJeKgME8qSxAv2jeelOXM9TqGtPM4FAnjRWIFAPEAiUAgKBUkAgUAoIBEoBgUApIBAoxf8DbNCIITHalwIAAAAASUVORK5CYII=" /><!-- --></p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="fu">plot</span>(stat_results, <span class="at">plot_type =</span> <span class="st">&quot;volcano&quot;</span>)</span></code></pre></div>
<p><img src="data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAkAAAAHgCAMAAAB6sCJ3AAABOFBMVEUAAAAAADoAAGYAOmYAOpAAZpAAZrYA/wAZGUgZGXEZSJcZcboaGhozMzM6AAA6ADo6AGY6OpA6kNtIGRlIGUhIGXFISJdIl91NTU1NTW5NTY5NbqtNjshmAABmADpmAGZmOjpmOpBmZmZmtv9uTU1uTW5uTY5ubo5ubqtuq+RxGUhxGXFxSJdxcXFxuv+OTU2OTW6OTY6Obk2OyP+QOgCQOjqQOmaQkGaQtpCQ27aQ2/+XSBmXSEiXSHGXupeX3bqX3f+rbk2rbm6ryKur5OSr5P+2ZgC2tma225C2//+6cRm6///Ijk3I///bkDrb/7bb/9vb///dl0jd/7rd/93d///kq27k///r6+v/AAD/tmb/unH/yI7/25D/29v/3Zf/5Kv//7b//7r//8j//9v//93//+T///+rYEZLAAAACXBIWXMAAA7DAAAOwwHHb6hkAAAgAElEQVR4nO2diYPctnXGKdcV2saWHaW1lWPTpnIO90oay4mjXnEab6vE3tZuI8+urVhbK8b//x+UIG7wgQQIYgjOfJ80uzzeYN4MfvvwcJDTcQgqULe1A9C+BYCgIgEgqEgACCoSAIKKBICgIgEgqEgACCoSAIKKBICgIgEgqEgACCoSAIKKBICgIgEgqEgACCoSAIKKBICgIgEgqEgACCoSAIKKBICgIgEgqEgACCoSAIKKBICgIgEgqEgACCoSAIKKBICgIgEgqEgACCpSKkCvQm2pKhUZSgaoqhdQrpqpDwC0TzVTHwBon2qmPgDQPtVMfQCgfaqZ+gBA+1Qz9QGA9qlm6gMA7VPN1MfJA/R/f/NL6vAX33il1z8c25vV1Ex9nDxAn/39m9ThL/7it/2PPyfh2oOaqY9TB+gP//xvf/e7/vcXb/yu3/7lZ6+88qe/FccHgPoDG7u3WM3Ux6kD1IPzG9FQCVa+eON//va3/JOvDceHCPTG7zZ2b7GaqY9TB+iTN/lnX1Mbn7xpEyKZA/0ZACrViQP0h396RTVaX7zxvyIIfeOVPxkYkk3YP+02i26mPk4coKGRUm3Yv8oG67Mh7AwAyTO7VDP1ceIAfSIIUW3YK28O8DgARbr4e1Az9XHaAP3hX0SckZwMffbfmF4YxoHW0WkDdLpqpj4A0D7VTH0AoH2qmfoAQPtUM/UBgPapZuoDAO1TzdQHrgvbqapSkaGyCHTIe7FM88rF79p5AFTBvilzAOQLALXlPACqYN+UOQDyBYDach4AVbBvyhwA+QJAbTkPgCrYN2UOgHwBoLacB0AV7JsyB0C+AFBbzgOgCvZNmQMgXwCoLecBUAX7pswBkK8TBIgxVrF0AOTr9ABivTmrVjoACgSAmnIeANWwB0Ab6HwBQg60ipIBOkAtaX8AkUd3HYF27TwAqmDflDkA8gWA2nIeAFWwb8ocAPk6FYCcrtf+nHcEgCrYJ5gzbgd/due8KwBUwR4AbSAAVMkZAOTrRABCDrS2zg2go5kDIF8AqC3nAVAF+6bMAZAvANSW8wCogn1T5gDIFwBqy3kAVMG+KXMA5AsAteU8AKpg35Q5APIFgNpyHgBVsG/KHAD5AkBtOQ+AKtg3ZQ6AfAGgtpwHQBXsmzIHQL4AUFvOA6AK9k2ZAyBfAKgt5wFQBfumzAGQLwDUlvMAqIJ9U+YAyBcAast5AFTBvilzAOQLALXlPACqYN+UOQDq9dX77+hNANSW8/sA6OkDALTUHABx/vyHPwVAS80BEP/qV7+WTdirQlvflhTytAeAnr59ojmQukPHPp1X2gFAz3/06WkCxNRjl85r7QCgpw+E3lZ7AKgN57V2ABA/1W484xGApr/7oA3ntQBQBfvCHIjxyW9facR5pX0A5OiUAIqYMw6A8gWAjBgHQPkCQFbIgRYIAFUyB0C+AFBbzgOgCvZNmQMgXwCIMJ9MmgCQJwA0Nmd8qtsGgDwBIABECgAtN2ccAAGgEnPkQAComjkA8gWA2nIeAFWwb8ocAPkCQG05D4Aq2DdlDoB8AaC2nAdAFeybMgdAvgBQW86fBUDX19c55rnFt20OgHwtAOhaPRLNc4tv3BwA+QJAbTkPgFLUFBFtOX8OACEHqmh+FgCVmrdFRFvOA6AK9lPmxMqLA314JWcAkK+dA8T4eO3XYfZawhJnAJAvAASASAGg8PBKzgAgXzsHCDlQLZ0LQLR5Bj+NOQ+AKthnmzOe3oI15vz+ANr6tqQ1xNRjh9ofQORRRKDV7E8+ApFHdw4QcqBynTBAc3Q07fycAFAFe9+c8Zn2qWXnZwWAKtgDoA0EgCo5A4B87Q8g5EBH0QkDtK05APIFgNpyHgBVsG/KHAD5AkBtOQ+AKtg3ZQ6AfJ00QEkzGm05D4Aq2C81ZzxlTrUt5wFQBXsAtIEAEAAqEgDiyIFKBIBojZlqy3kAVMF+RXPGR61aW84DoAr2AGgDASBSjAOgNJ0HQGSWjBxoDZ0FQIxT/fSdOE8LAFWwB0AbCABVcgYA+do1QDM5UNrVYQCI1HkANGnOeNL1qW05D4CmJe/PCYCiAkCTUncIBkBRAaBJHRUg5EAlAkCVzAGQr/3kQMl33ABAa6hRgJYXz3gsownJatD5dAGgCvbTAI1ONOh8uvYA0OcPHnzrI70DgAAQqThAz3/wEX/6Hb23G4CiORDjAKiCppswAZHUfgCKCjlQDU0DJCPQq0Jb35YU8rQLgJ6/9c0P9PYJRKDjmiMCCX35E03QmQDkN3PeXlvOtwnQVTfovj3y4Ttq4zwAYtxNtP29tpxvEKDbC03OVXfncd+L//anZxeBGAdAmTIA3f7lE3t02Hn64MFp5kCqZQrNmTzO7AEOgGZ1ciPRMdl8hqlHYM64YWj8nHxnAJCvvQPEuIkmepMCaD1nzhOgPot+ePXyx4QdAAJApDyALl/+r4uHLx7dJexOCKBRDqT2Z9aBACBSLkC3Fw/7//zmpSdju30DxJjNbsyGNmc8DD35F7JmebOGOQCqYB81Z+rhd7SiAI0OLHHmHAHiV6IJu724T9idBECM5wDkhq18Z84SIH4TDERbnQZATD1887DFkuaebb4zmwD04pGcS3ioDzx77fHw27YqYrZBjBPrM6U6j268SZPdoZ4Zc7ZLgIK//hCgF49EH/uqJwwALTEXZLBEc3aKAF3KMZqrlz+uAdDthQyAJ5dE21YqJ6nZYQ5kARKN2V0JUF+td34u61T0kZSevfZz2dY9uzekLc9ee+/esC/M33v9yVACRUKgcQS6/S6FZgMAyUt9FhXPONGv2g39lGYAEoN54tEDJDpFtxcSBSfsPLvXB6Orl54MTPW/7b40H4YD6UFlT0QTdkM9a3uA1MWGi4pn/FwAUutxhiar/9ETM2xeKYBeNyHl2b2HA1C/F5Xd/9b72nz47USsmCiA2mzCVgHIaZNyvvJ7uvRV7NeNQEMQEFi89niIIoocNwK9prtiN0O/TO9r8/HaMFoEQJerRCCqwdkMIGe2wgxI9+ZsbLGo9DXsjwKQjih9mqKBub3oO/XS0gdovvUaRCTRd9bIgcjq3iYHcqMOd6ZUPYCcM3mlr2W/MkCiEp0mTLcqCour4YxsssSRGxuBtPkNicFYtbrxFQBaaM64G3YCgKgz6zjTUBJ91yTR4TiQAEYEoHt3nIikk+jeMoGiMwDI6Y67HXM9JqTO8JMCKNKN73XpjkSLH/2BO7+4eGj2tbkoISEK2SWtahBorXGg1XOgpeYDKKOpeGkudogzaziz67kwsiNO6wxGoiVBzo41F9ts/Iw1nNkrQKLVopeE0TplgIKul9NmDQcOPCHuLHZmrwANs63p/PgADaPaJzOVwbgz+CMpUos51M+8gaBzAShTLkB9Cvbi0UN69HHHAMnIY1Mhpo/mEQSASAUrEvnl/VanMrLNmXwMv5gGRq/U6DcO6fnPAmfOFaCru61OZeSb2+WHJgY5ZwDQKvKvyhjoIQex9wiQFON6MNFvsdCErSIPIDEOdbnOVMYa5usNJNpFhuqI86uSM2cJ0IR2CxDjZoEzc4/05ohAK+h8ADIJD/MeyRQBIFJ+Eh1f/XGqALknVnXmHAEaBiEjS9B2C5DuizlTYnpkmgOgcoVN2OVK64HWMF+xeJ1Dm4SaqRyIcQBUJGpFIjkOtPVtSYvE7D8h9fOgNrf2bomaBegyci3HthFodjEiVbw7cKj/qSFpO0PPkrvziECk/IHE+BKiTQGaXw5NFM+4WUxmZlNlA2Znx+w0a27pq9sfAaDhwgv3ygytq/vqai7HQubDLx4NB77+sTr/9dEYcziVkedw8wApXrilx0fK76BNkrQvgKiY/eyeaFzGAA1Y3Pz1fddCTEeIXrlga/gxnJe7nnYxDlQOkJoN4zbmuBwpgNQzVna+jvkMQOQn9uz1/7gr8bi96ORlX90f/dVjftnHjRfv/uJ7H1sLecvVfkPA9eLdx+q8fyvWQSFAsas59pkDMQcgZjpdo8Mq8DB+6gA96VkRAF3eF7Pmw9KLO48HKHpQ5Dn5U67I6BswAY+ASJ7nA2yemgIo5KS8eDfU6BTarBCyuZAy5icP0O33PhY/+3ymx0aQ0wMytGB923Rz11rcyDWJPS5XshmT54k2rCWARu96/dl4uSOZkXNhem8wPfEcSOQ29wUkF/JiVNk8iWgzXBP90hNjIQESKfSzr/+3aMEeqZWq5wqQaba4A5A6P56nnym9Sq/tGL0wcc+Nd9/TGQ53ItAQhS4fBhbD7s/+055PACjP4d0AZBnSKQ8z5sqAz7RhpvRJqzlv1jJfCpDsaKl1pzYHuhKpzc1dayHuxiDnRq9E90udn82BrrrYbNhOcyAhN1F2G7LzBEihMQz49b/++F3RC3vxM3Hm9rvvGYvhlgsDCeJiQ33+8Uwv7Er27ba+yabh6EAeJfbixetwo+rc9tpt38tsnTpAlARUxPBgTNPjQGog8ahrogkQbEt2II+O9+LFM27aLzbsO72yg0HBhSnB+b3mQKHMxctjLCJKG4k+JkAUCDUAMqHIpEAeQDwhpuyrF3Y0eU3YjWwYj9iEHQkg09VyhoIsUgCoQNRNNqmLU48IkGjWZMs2yoFsg5eRA+kmyx9UdEcTGQdAS7V1N56epFBcHYLzM3NiZASS6Lj06GNMnvbS6Gnn05KfuDcrmgOgSRmAEhquqeINMNyFh/EBnQMzAKU5L5+XqnMDaPyVl552CZCZ73LYYbo/b8+4VMSDDAAi5X3prhpCvCKWlY0cHtqWah+SzoGGxCc8nFy833LZBGg4xQ9mQMh9RpQRAEQqXJEYu7Vr6LAMBrWXc1wnLuSIFK/THduNd8+xYBBoOMonAEIORGhpDlQIUOICn8SVZKPitRQ+bJQqs2EcaDwI5O3MlZ7rzZrmZw5Q6hLDYoBUgPHaL+eI24c3J5NLz/VmTfP2AHIGgZLGgYpyoOQ1qvMTYFTxVn7yY4Fh6sHdweog9gQsASBSDkDOLPx4NmPdXtiiRc6zz4p047kLkJ3N8MaiCYDCAwCI1DYALVvknA2QCTTBOLS/qWy5mwoR8xsnAFDfyqTfvzdRbQ4k0vYLAfK48Vs1bmc13BaLqefuGKA+DRkdEytUrzJuwJok52XU97NsuKBszt6NW0QMi48DOe2WYYcbgBgnW68950AdJ2KDGB4mLissk30VfUm8+G6FsbYBKN7SUdGI7sZbjrgZOXSZigJElJSj9gAyF5iuKSIHaucmm8FExty8Kt0Ls2GHMzcEMffwdAdMD1xnqT2AxDLoMwZoflqMyIH8ER8nIGm0tN2MJ2xnAJE5UN0IZL5nrqEmbAKglBzIDPy44caJRyr4ENp9BCJVOQca1uE3lkR7WTPP/cZCjxwHG48nax30woKy9pUDkRI34a3ZC5vUhr0wTVH2NxYy01iRDPnmjFto3O1C52uZYxwo2T6MPHGOyByICDtE+DkPgGpoAqDnbz148I7eaQagiZYsspwjorDjxYOR6FWcr2a+A4C+/MkH/Pn3P1B7211YOE6eYyFo3IRFoo9q2UbWk5kOACIV/8rLz7/T//hQh6AjAhRe1hPMyPPUCGQ4GWHEfZCYfcZEtgyASLlTGePxHxGFemeFjncD0mv1kDu9/LOjIzGZm7KOdFA3ZxX3Z5UH9DMO+nDrahEgMxJk9NX7b+vNjSKQCD5+i5UfgUatmGrbvEub1TO4nF1d7vxi+xOIQCN9+WPDz1Y5EAFQei/Ma6b8BMhbsaE37dxZgfNL7U8QoOdvvWN3NuqFuQAN5EwNJwYDO2aynRD3weI68ZF7Kzlf07w9gEYLyjx+NuvG24Aj0UkGiPEIO5oZPTfB3Ekx9eAERgCIVBygpw+ENuiFRezzAbJTpkQS5GY+BhcWPFZzfnXzFgHKW1Q/qOKHJCIPAdDUhMYYoIkAZDhyww3jAChTrU5lDLj49tdaacW7OQ7Vgqlr4/noSRwAZagBgEgkCIB0AErrhU2FHw1QxnoOABQRfX+g8Zx/PYDorCYOUFoOxHiMIS6Heuy2KwKcSecndIYAmZtsvng0Img5QE7EyABonAPxTICiqY+35TPDeGwU6BQAenaPvvNByZf1RO6RuOJ1YS4fOQCR9tNJkAcQ8yMQZ0FKpA4Es/D8VACiQulAxnhNa9GX9TQAULBY3t2h7JN7YWTkcY+brzoYA0R9/LsCiHHiL2EASAQV9WU9N7K3XfRlPZH7RI+/8KAiQDHrojogxn9GzZfc5O5HzcxcBuOBTgQgcd9weaP64Q71d3nZl/UEvbAbuSb6KuEGU4PWyIFC4+t1ACJx0XMYw8YwEq3iki/GTxOge7J6BTLirvOyMSv7sp4WuvGukiJQrBWb6YUpdsxhAZBpxDyKGNcfv3N8VwDFcyCuv3TlsQCqjxRlX9bTHEDjCDTKm6NZd5gDcWoWI4hHTI5Zc/8PVn/87vF9AURJA6QikNjs6Sn7sh7qy1bI+95vFoHGXfdkgGxffcSRH5iivS/3+OkApHIgEXp0KrT4y3qO8mUr9JenzNjq7YUA2WSHIEY3YWY6TD9CucfD5WrxIUfSfk7HBEj1wi51L6zgy3qO8V0Z4RrnVImBxGsL0PRd6sPip6OOG34Yl+GKLNI5Hsy08fiIEWU/q+1Goou+rKd9gHQONH9h6hKAJDmMczV7n1j6cQGi/FpzKqPky3qO0YQVAmRLmbs0NZoDJQUh/TNG0WYAkS/V5lxYpSQ6JwcinzfsqH8TCoufizvMYMYNQHIzofTpcEV4M6PTACiuzVYkOlfGT8yjhsVHo4+HlR4Hchqy4cdqzq9izjgAWmYu7HX7NzMRHxTPeDT+UMftMyaapu268VSwaw+gqS8L400AlJEDMa7TmclmzEYhC020adr/OFAVtRuBxAWoBpzrFIiSItCInomoYyyync+3B0AJcswpEMibIGp4NENpyzlS4DnY9kuHKr88eUptACBaIUDjhRxSa99onI/zGf+Ys+dujq5UjXmTEoDUZBjX20Eckqc4AJpSKwA5rdTIYgFASejoCMQGThgDQAt0dICcdMY5QxyTOZDedI7y6/Bg6A0LZ90nSbLpdpgJMeRAszo2QCYpJu65YbIdeUgA5Fra05GpDe1NvO/FiT0dfpgmRRUSZkQAiNSxk2gq/HA3OQ767Xb40ObQ8tS4EAeglLkwZ0204skWZU4tfq8AyFdlgGzTZA0UPB5A1zYQjQuZj0Be+DnoCOQkQbaEUUoNgEgdvRs/ypalzDEbfZwHtwev9TKh6RwoWfoJHjJzALHp1m5knyAAlCAiKdayUNlxn+BieL0THQvyAZrFKJhedxstNp0DMe62b/4eZZ8iAJSgKfPRumdxZapdk3htjl5Hr483TVhSG+a2WCYZ8o5EnfdMAVCSjniTTaFrdTNNecNNddtNccAcn7jRZuzeml74UQ+zeZD33UwSO7h34vT3jqT9AUQeLfkrm5rVuvY76/qn/y9afOookOmuqfFCR9MRKDhNGIdvdl4nH4HIowUf0tQSVdtUuTdG1E1Z9AL53CSa65/yGU5JLNhf8F4BkK86AEWCkMyA7KCz0/EyOdBkBMqTCiEOM4yP0xoARGpLgGIrxEycOfiYmU5Z5ImqeJbaDaMjEoss8ABApLbOgQgOrnUOpJotY3Xt9sHmAVoQhlQwUkeK3isA8lWjGx8ZkdYNmLMMegTQVDeeZY4k2thjAOJjfgAQrU3HgcYcKFo0QLYLpu2j4YeHOVAKRTz8qQEqf68AyNcxr403Y9EHdxRRnteD0VPFs0VtmPOshc4X2QOgBKUA5Kzp0GuiPYOJRa2LemHc3yxxvsgeACUowdyLL0Hs0T9NZz5SfELrRVnE5iTSnS+xB0AJmjYPUmbH3szgG3rIDHzpOJDixwGICkUAiFQ7ALmzFlzidHDPUAD53bGs9UAhPMwBSIejkvcKgHwdBSA7dyEZoQHqbF/NnxSbmMrouo7ixvnJbA5ETY4BoIhaAsjgIDnpOhcgyVR/sP8hdvqt3rojAeLM8tLpn/KqW0mTwxRXR2XR8tkMACWrHYBsDjQA1ImHqlTVUIk9wc212BQcRQHqmMTChp1hf/g3nBOPA1Mm/a+hQP1h2O6YhQoARdQQQP6yVsGJU6lySyGkAOoiOdBgx0zAkTGn0wgNB4c9roOSKpDbEMT0K9rXB0CkWgKIy5bJbNhKldBwVeOyvjsvPnjFO198pmFR/HTOnj3oAOSVCYDm1QZAptpslVmA1G8trhHqQu8pgGIKbZwCvZhHAhSSm/dmVzEHQKETnd5StaOh4brBCrjg6hxV/AiPaZRIgOUumQMFRplvdh1zABQ6oQGS/7jlRpNjM6DOwBQpPo2cgxd8usCTCecBkKN2AOo6N+gMWwe9M4o98kQUoFSGdHk6ByKoJJwP2rnIMwCQp+o5kBt1dHuisuWuCyGy2TRdfBI1zm9uo1qK865ZNBwBIE/Ve2EaILd2TI3qeiZAoItPDT12g0+wMOU8AErTPEBTf72EOeWIyZVtkZogt8YVTzq3HhefiI5OfoIXyXMeAKVpFqCUzHLyQ7K1aAryKpyHO1GA5jth3N3kekzJnqJ8iyZwyIFStCZA0T9zA08AkMOPOxitz4bezAYgj6XAnHojwyEMJJLaACCTEVO+dGOAgkjUeTmvG4QO7rNyGNJByB1DGDkGgEhtkAPZ3IPwxcuBDhYXL2TYzloZQHoAyGvDAFCWNuiFzQFk5y4Opm/dhY2WrvoxQNkByGWTc/IvoZvIgabfbC3zcwbI4EPWk8eQ2/sKRoKUvVvbuQD5Y0HWH4Kg/PcKgHytPA7k1/zIpaCLRVV8tHhyuGgGJC+URbI5AESqmYFEqwAgL1joKudUuFiaRHedVw4AylGLAJka7aipMLfW/drOBOhgtuzYQFjk8vcKgHwdDyCXGJtP+zWttzsCoOwWzEa7ySwaAJGaBOj5Dz7Sm5sAdAi7X6OK79YASL7G1DhEyiDXoje72HwXAH3+4FvVAIrXlgSIa4C4V81RlmzxuQCZJEtm0nGf8t4rAOIffvPfq0WgiRrR1WlDg2WCAsiWtSyJDltI0qe4uwlvtob5HgAyTdirQuveZLRTD7nTdXZr0MGp4MFyOHLwThx8C7fwfDkv0Xll2RLX/QBKtSuAhCpGILvd2UdnWhUv7IzDC+/CCLRgMsyZW+vchqxTKwQ4pjJi2gqgzlZTAJA61XU2AXI7X2OATMOT14TZsv2Bbj72DADFtQ1AXk5BA8S9avajRXjEKz4vieY6grlghZ4BoLgaAMgdBLZMeFU6osKbsPCLz8HHLaJzX91xR3sIgEi1ANDIIdWTt40KgZBHl1v8gnEg73lyp/Oi4YL3CoB8rZ8DxYw6Z2jGUhJAMwVQDjn6wTVA3GbNOg9b9l4BkK+jfm+8/us/EHU9pqgAIFOEeiL3APLHFgEQqaYBmux+OVz5xefjIxlyWRoe/sAiACJ1VIC6eMMV2tlOvJ2Wj1R9MUCaHu7xGmZqAIjUMQESPeGk1xszMlX5o0YmnRtvQw1aOlB2LvAAiNR2AHXxcOTMpyaB5HmTjE9Yuur5cZsJeR4CIFKbATQ1P2lyoBR+bD0f5HMX0eOsDdEABR4CIFKb5UCTE9yyXhObMn8qIxsfN9uxrzP2EACR2qwXNgnQYJBc/+UA+a1V58ylLH6vAMjX+t34LsiBwn2XhVFGFCJki88Bx/BDJmQAKEXNjAO51WVq1ZvinM2jF0Yg3YUfecTRhM2rFYDcOtRIeCPRU/hw+dSlTVgXBhvlBgdA82oEIKfzbAcObe8oAQG+ACDuPcZOoRs/r3YAMkHAVmg4FzZFAl8agcarOHKdX8UeACUoYi5rz1ShTW4P43n5KAa8oAkDQMvVAEB62NBbhZNZ9wagJGPy0FwbBoBIHRGg2B+5XjthR+90tR74XAfemCpvukz+HPO5bhgAInUUgDrVxTqQr2YikKkwv2bTQOAKoKyneUUAoCU6BkB6iikCkJ244G446oYFZYlAaG8WcdP5y4oCv/Pea+5ns9QcAPkGnTsYk0+BHYleAJB8PuVahxxoVscDaGL9hrv4z6lVf2ueoEUAdba5nH4XAIjU8XKgCXOyA5Y5JMiXA2Tm4w1EFE0AiFQD3XhuJ7+4XtNluUjDqQAg/ZrObAq5UgAAkWoDICVdmQoYfpgnx+GgMALZRdgAKEPtAdQFLUoyP5kACTy57YHZ4x0AytB2AHWxJTimJz8MJNYCyF/1yE0j1tmMrOi9AiBfx16ROFTrIXkqTAOUzI/3VLuozL3Qdeq9UoRN2c8LACXIMe9iYy/2/ABQ8tgyXwaQ2/cyDejse52hf2SfIACUIGuuwaB90tMaXUZnni8EqNOLQdKdB0CONgMoEoO82j1kzG3x5QB1U8M/5HsFQEZtAGSrkOuevNjIIYJnA3TQM/3OMNTE54EciNSGOVBnZsBtzVmAuqEXVg0gbtou64cH0AgR9MJIJQO09m1G5c1b1LZ8dPKADRGHQ5d8y1VdarL18Ks7OHeMdW/3Orr1a2PaH0Dk0ZJuvOk+6z99P28W24ecccH8HMjL0qVPNgCNPhtEIFLbDiR2zrZbZ8cByG/IfFcBUKI2nMrogkqz0xe6c51zy3CeBZAwPNgB6DFAyIES1dJcmK5b06oczDjN2gB5T7TdvtRu/ApvttAcAJGyI8OSCBsuZikail/AT1BGYjd+jTdbZA6ArAfOn703P55JxBKAuE3eedY40NI3u5o5AHIdcPo+nZ7GkPZmwHF9gGR8U6wCoKVqCyBTnZ1uk9InMzhf8jU96qnmZfOcz32z65kDINcBekYsd1QnAyC13tHyauGJQTQJV+qbXc8cAFkPiKrpZGfMTaJXA8hMXhxMs9XZ/DnWjKnVbaVvdjVzAOR5EI7iyUYlMwilRyBvuVrnDWICoEw1CvUrwKEAAAWWSURBVJCqZs5T408yQGrNqnlRfTAGUKdn6gEQqTYBcpPpLrkNS+mFhcC4k/9q1+dHH+qQA5HaHiCqatShQ1DvcwClGNmEWRMzyYYeqEYvLKIGAIrJiRdJBCXnQKos5c306A8AmlFzAJlwILtVGVlQFkCON91k49TpaAWASLUGkJPd+nNhawBkR3xcgLyXp2DyeVvxzRaYA6CIKIBSA8ucuV780+mZtolXL36vAMjX8QFKngQzSk2ibekTr176XgGQr+PnQMbeiyClAHWdHtehbncFgLLVHEAj+0QqUgAyCx5jAE0l1ACIVHMAdUEEymjGEiIQd26+EbvhXoHzBfYAKEF5OZAGqLPDf2UAmUna4YUqjywDIF8NAJS5vMMNOxYgB5xjtL/VzAFQRCOAdGfMGZleBJEzrVXN+RJ7AJSgZb0wd8VQAUKao5rOH694ALTMfik31I2jANAa2hlARQEoLOjozq9oDoAW2i9rw8zFHk456Mavor0BVBCC/GIA0DraIUALZsk4DxssALSS9gpQRtih78ZIHKrv/HrmAGixfZc6MN253+FS7g0AIrU/gHIasDW9AUCkJj7jL3/84Nuf6p2GAFLSQSYWfGabKAC0huKf8lfvv8OffkfvtQeQ0ARB0TvOL/UGAJGKf8pf/uNH/PkPPlJ7bQIUjz98vnXe3PkS8x0A9PxHn/Ivf/JBv/Wq0Na3JZ2UC485sq1LlbUDgD7/tgZIqNEIZJSUNhd5gwhEKiUCCbUOECcHe9YrHQBFtPMcqF1zAPTV+2833wtr2BwAtT4O1Lg5APIFgNpyHgBVsG/KHAD5AkBtOQ+AKtg3ZQ6AfAGgtpwHQBXsmzIHQL4AUFvOA6AK9k2ZAyBfAKgt5wFQBfumzAGQLwDUlvP7AwhqS1WpyFDmxVFlqvyu6xa/a+frCQC1UToAStGu62DXztfTUQGCTk8ACCoSAIKKBICgIgEgqEjHBOj5Ww8evFOrcO8SgNVV1XUhcSeCXeqIAImrFJ9//4N5wyXybwWxtqq6PuhpXT7r6YgAfS7q98NKn5N/GeTaquq60PMf/hQApcheK72y/Auxa6hm6V/96tdowlIkrnatI/9WEBVUz/VeT99GDjStDx88+I5IdKtVQu0IVNH1wXkAlKDnb9X7kOrmQFVdFxm0UEVCK+qIAFWtBP9WEGurLj9CiEDzkn9ntT6nquNAdV0XAkDQeQoAQUUCQFCRABBUJAAEFQkAQUXaIUAvHslbQj/UB5699nj4ffPSE89QH08t9iH1hNuLruvLvf1uTllnpF0CdN8/sA5AV3epJ9xe3OX88uWP+U3/gMYCQEoixBBPuLnzWBbUB6h8V89AuwZINGZ3Zf32Lc2dn2uAzPGfy5bu2b3+9/1+/717w74wfu/1J4Odfs5VH2EkKMNzjY3QABZCEKk9A/Ti0d3h0dfv7cX9vsoVDOLY7cXDnpu+zq9eeiK2xW+7L42F3QCOLrQvSJfpFSiasMwG8Wy0S4CGHPq+bLL6H33VDptXLzkBQ/y+93DY+f3H8qDe18bD74Eu9Uufu1HnVIFXohnThpCvXQKkItCNCgz9/yGOPFMtzo0Lkmx+euDuPNb72viqUyRyByBVplPglezvjVIvSOjEABJM3A8Bur2481jZ+QA5WU0coMs7j/2XhVztGiBRtU4TdhM2YQqYAYobG4FMM3XHZjUGIFWmKVC3i2jCaO0ZIC+Jvusn0eq4geLZPQuQTaLF+I6i6JJOom3qjCSa1J4Bmu/Gqxzoss+AfnHx0OxrY2Gno9AN0Y3vba7MqPcVuvGUdgjQOgqHdai5CtcGA4m0zhEg0WoNY0Ceru5O2mAckdY5AsRFuxTyE4aYwAaTqRGdJUDQegJAUJEAEFQkAAQVCQBBRQJAUJEAEFSk/wdIoQsUuJaGcAAAAABJRU5ErkJggg==" /><!-- --></p>
</div>
<div id="p-value-adjustment-methods" class="section level2">
<h2><span class="math inline">\(p\)</span>-value Adjustment Methods</h2>
<p>If several comparisons are being made it is a good idea to adjust the
<span class="math inline">\(p\)</span>-values in order to control the
false positive rate. Currently <span class="math inline">\(p\)</span>-value adjustments are made at the
biomolecule level (i.e., adjustments are made for the number of groups
being compared). The available <span class="math inline">\(p\)</span>-value adjustments are described in the
following table. The “Appropriate Comparison” column refers to which
correction method is preferred based on the type of comparisons being
made, i.e, all pairwise or case-to-control. Check marks in the “ANOVA”
and “IMD” columns indicate if that correction method is appropriate for
ANOVA or IMD comparisons.</p>
<table>
<thead>
<tr class="header">
<th>Method Name</th>
<th>Appropriate Comparison</th>
<th>ANOVA</th>
<th>IMD</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>Bonferroni</td>
<td>Both</td>
<td><span class="math inline">\(\sqrt{}\)</span></td>
<td><span class="math inline">\(\sqrt{}\)</span></td>
</tr>
<tr class="even">
<td>Dunnett</td>
<td>Case-vs-control</td>
<td><span class="math inline">\(\sqrt{}\)</span></td>
<td></td>
</tr>
<tr class="odd">
<td>Holm</td>
<td>Both</td>
<td><span class="math inline">\(\sqrt{}\)</span></td>
<td><span class="math inline">\(\sqrt{}\)</span></td>
</tr>
<tr class="even">
<td>Tukey</td>
<td>All pairwise</td>
<td><span class="math inline">\(\sqrt{}\)</span></td>
<td></td>
</tr>
</tbody>
</table>
</div>
<div id="false-discovery-rate-methods" class="section level2">
<h2>False Discovery Rate Methods</h2>
<p>Coming soon!</p>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="unnumbered">References</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-rcpp_book" class="csl-entry">
Eddelbuettel, Dirk. 2013. <em>Seamless <span>R</span> and
<span>C++</span> Integration with <span>Rcpp</span></em>. New York:
Springer.
</div>
<div id="ref-webb2010combined" class="csl-entry">
Webb-Robertson, Bobbie-Jo M, Lee Ann McCue, Katrina M Waters, Melissa M
Matzke, Jon M Jacobs, Thomas O Metz, Susan M Varnum, and Joel G Pounds.
2010. <span>“Combined Statistical Analyses of Peptide Intensities and
Peptide Occurrences Improves Identification of Significant Peptides from
MS-Based Proteomics Data.”</span> <em>Journal of Proteome Research</em>
9 (11): 5748–56.
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
